{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>12012</td>\n",
       "      <td>12213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00008836\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00000001\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>183635</td>\n",
       "      <td>184238</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00006031\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>183635</td>\n",
       "      <td>184428</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00008837\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>597422</td>\n",
       "      <td>597636</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00021307\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source  type   start     end score strand phase  \\\n",
       "0  chr1  StringTie  exon   12012   12213     .      +     .   \n",
       "1  chr1  StringTie  exon   69091   70008     .      +     .   \n",
       "2  chr1  StringTie  exon  183635  184238     .      +     .   \n",
       "3  chr1  StringTie  exon  183635  184428     .      +     .   \n",
       "4  chr1  StringTie  exon  597422  597636     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  transcript_id \"TCONS_00008836\"; gene_id \"XLOC_...  \n",
       "1  transcript_id \"TCONS_00000001\"; gene_id \"XLOC_...  \n",
       "2  transcript_id \"TCONS_00006031\"; gene_id \"XLOC_...  \n",
       "3  transcript_id \"TCONS_00008837\"; gene_id \"XLOC_...  \n",
       "4  transcript_id \"TCONS_00021307\"; gene_id \"XLOC_...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import argparse\n",
    "import sys\n",
    "import os\n",
    "\n",
    "infp=\"./hisat_to_ref.annotated.gtf\"\n",
    "\n",
    "gff3cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "df=pd.read_csv(infp,sep=\"\\t\",names=gff3cols,comment=\"#\")\n",
    "df=df[df[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>12012</td>\n",
       "      <td>12213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00008836\"; gene_id \"XLOC_...</td>\n",
       "      <td>TCONS_00008836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00000001\"; gene_id \"XLOC_...</td>\n",
       "      <td>TCONS_00000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>183635</td>\n",
       "      <td>184238</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00006031\"; gene_id \"XLOC_...</td>\n",
       "      <td>TCONS_00006031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>183635</td>\n",
       "      <td>184428</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00008837\"; gene_id \"XLOC_...</td>\n",
       "      <td>TCONS_00008837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>exon</td>\n",
       "      <td>597422</td>\n",
       "      <td>597636</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00021307\"; gene_id \"XLOC_...</td>\n",
       "      <td>TCONS_00021307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source  type   start     end score strand phase  \\\n",
       "0  chr1  StringTie  exon   12012   12213     .      +     .   \n",
       "1  chr1  StringTie  exon   69091   70008     .      +     .   \n",
       "2  chr1  StringTie  exon  183635  184238     .      +     .   \n",
       "3  chr1  StringTie  exon  183635  184428     .      +     .   \n",
       "4  chr1  StringTie  exon  597422  597636     .      +     .   \n",
       "\n",
       "                                          attributes          parent  \n",
       "0  transcript_id \"TCONS_00008836\"; gene_id \"XLOC_...  TCONS_00008836  \n",
       "1  transcript_id \"TCONS_00000001\"; gene_id \"XLOC_...  TCONS_00000001  \n",
       "2  transcript_id \"TCONS_00006031\"; gene_id \"XLOC_...  TCONS_00006031  \n",
       "3  transcript_id \"TCONS_00008837\"; gene_id \"XLOC_...  TCONS_00008837  \n",
       "4  transcript_id \"TCONS_00021307\"; gene_id \"XLOC_...  TCONS_00021307  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"parent\"]=df[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\";\",expand=True)[0]\n",
    "# df[\"parent\"]=df[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.sort_values(by=[\"parent\",\"seqid\",\"strand\",\"start\",\"end\"],ascending=True,inplace=True)\n",
    "df[\"next_parent\"]=df[\"parent\"].shift(-1)\n",
    "df[\"next_start\"]=df[\"start\"].shift(-1)\n",
    "df.dropna(inplace=True)\n",
    "df[\"next_start\"]=df.next_start.astype(int)\n",
    "df=df[df[\"parent\"]==df[\"next_parent\"]].reset_index(drop=True)\n",
    "\n",
    "df[\"start\"]=df[\"end\"]\n",
    "df[\"end\"]=df[\"next_start\"]\n",
    "df[\"type\"]=\"intron\"\n",
    "df.drop_duplicates([\"seqid\",\"strand\",\"start\",\"end\"],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>735549</td>\n",
       "      <td>738961</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00000005\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>818202</td>\n",
       "      <td>818723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00000007\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>827775</td>\n",
       "      <td>829003</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00000009\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>829104</td>\n",
       "      <td>841200</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00000009\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>StringTie</td>\n",
       "      <td>intron</td>\n",
       "      <td>829104</td>\n",
       "      <td>847654</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"TCONS_00000010\"; gene_id \"XLOC_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source    type   start     end score strand phase  \\\n",
       "0  chr1  StringTie  intron  735549  738961     .      +     .   \n",
       "1  chr1  StringTie  intron  818202  818723     .      +     .   \n",
       "2  chr1  StringTie  intron  827775  829003     .      +     .   \n",
       "3  chr1  StringTie  intron  829104  841200     .      +     .   \n",
       "5  chr1  StringTie  intron  829104  847654     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  transcript_id \"TCONS_00000005\"; gene_id \"XLOC_...  \n",
       "1  transcript_id \"TCONS_00000007\"; gene_id \"XLOC_...  \n",
       "2  transcript_id \"TCONS_00000009\"; gene_id \"XLOC_...  \n",
       "3  transcript_id \"TCONS_00000009\"; gene_id \"XLOC_...  \n",
       "5  transcript_id \"TCONS_00000010\"; gene_id \"XLOC_...  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.drop([\"next_start\",\"parent\",\"next_parent\"],axis=1,inplace=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df.to_csv(\"hisat_to_ref.intron.gff\",sep=\"\\t\",header=False,index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
