{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Curated Genomic</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>258511</td>\n",
       "      <td>259036</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-RPL23AP21;geneID=gene-RPL23AP21;gene_n...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Curated Genomic</td>\n",
       "      <td>exon</td>\n",
       "      <td>258511</td>\n",
       "      <td>259036</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene-RPL23AP21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>490756</td>\n",
       "      <td>495445</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-NR_028322.1;geneID=gene-LOC100132287;ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>490756</td>\n",
       "      <td>494898</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_028322.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>494992</td>\n",
       "      <td>495049</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_028322.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid           source        type   start     end score strand phase  \\\n",
       "0  chr1  Curated Genomic  pseudogene  258511  259036     .      -     .   \n",
       "1  chr1  Curated Genomic        exon  258511  259036     .      -     .   \n",
       "2  chr1       BestRefSeq     lnc_RNA  490756  495445     .      -     .   \n",
       "3  chr1       BestRefSeq        exon  490756  494898     .      -     .   \n",
       "4  chr1       BestRefSeq        exon  494992  495049     .      -     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=gene-RPL23AP21;geneID=gene-RPL23AP21;gene_n...  \n",
       "1                              Parent=gene-RPL23AP21  \n",
       "2  ID=rna-NR_028322.1;geneID=gene-LOC100132287;ge...  \n",
       "3                             Parent=rna-NR_028322.1  \n",
       "4                             Parent=rna-NR_028322.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61879\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sparrow/anaconda3/lib/python3.6/site-packages/pandas/io/parsers.py:702: UserWarning: Duplicate names specified. This will raise an error in the future.\n",
      "  return _read(filepath_or_buffer, kwds)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>refseqID</th>\n",
       "      <th>covlen</th>\n",
       "      <th>CDSlen</th>\n",
       "      <th>gbkey</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>gencodeID</th>\n",
       "      <th>covlen.1</th>\n",
       "      <th>CDSlen.1</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>gene-AKR1B10P1</td>\n",
       "      <td>1201</td>\n",
       "      <td>0</td>\n",
       "      <td>Gene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENST00000425765.1</td>\n",
       "      <td>942</td>\n",
       "      <td>0</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>gene-ALDOAP2</td>\n",
       "      <td>1377</td>\n",
       "      <td>0</td>\n",
       "      <td>Gene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENST00000398050.3</td>\n",
       "      <td>1076</td>\n",
       "      <td>0</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>gene-ATP5MC1P7</td>\n",
       "      <td>550</td>\n",
       "      <td>0</td>\n",
       "      <td>Gene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENST00000444398.1</td>\n",
       "      <td>400</td>\n",
       "      <td>0</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>gene-CCND3P1</td>\n",
       "      <td>628</td>\n",
       "      <td>0</td>\n",
       "      <td>Gene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENST00000438347.1</td>\n",
       "      <td>580</td>\n",
       "      <td>0</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>gene-CEP164P1</td>\n",
       "      <td>504</td>\n",
       "      <td>0</td>\n",
       "      <td>Gene</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>ENST00000455142.1</td>\n",
       "      <td>418</td>\n",
       "      <td>0</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome        refseqID  covlen  CDSlen gbkey gene_biotype  \\\n",
       "0      chr10  gene-AKR1B10P1    1201       0  Gene   pseudogene   \n",
       "1      chr10    gene-ALDOAP2    1377       0  Gene   pseudogene   \n",
       "2      chr10  gene-ATP5MC1P7     550       0  Gene   pseudogene   \n",
       "3      chr10    gene-CCND3P1     628       0  Gene   pseudogene   \n",
       "4      chr10   gene-CEP164P1     504       0  Gene   pseudogene   \n",
       "\n",
       "           gencodeID  covlen.1  CDSlen.1                   transcript_type  \\\n",
       "0  ENST00000425765.1       942         0              processed_pseudogene   \n",
       "1  ENST00000398050.3      1076         0              processed_pseudogene   \n",
       "2  ENST00000444398.1       400         0              processed_pseudogene   \n",
       "3  ENST00000438347.1       580         0              processed_pseudogene   \n",
       "4  ENST00000455142.1       418         0  transcribed_processed_pseudogene   \n",
       "\n",
       "                          gene_type  \n",
       "0              processed_pseudogene  \n",
       "1              processed_pseudogene  \n",
       "2              processed_pseudogene  \n",
       "3              processed_pseudogene  \n",
       "4  transcribed_processed_pseudogene  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this assumes that intersection has already been computed\n",
    "\n",
    "df = pd.read_csv(\"./prepAnnotation/refseq_vs_gencode_intersection.gff\",sep=\"\\t\",comment=\"#\",names=gff3Cols)\n",
    "display(df.head())\n",
    "df_tab_header = pd.read_csv(\"./prepAnnotation/refseq_dta_gencode_dta.tab.header\",sep=\"\\t\",comment=\"1\")\n",
    "df_tab_header.reset_index(inplace=True)\n",
    "tab_header = list(df_tab_header.iloc[0])\n",
    "df_tab = pd.read_csv(\"./prepAnnotation/refseq_dta_gencode_dta.tab\",sep=\"\\t\",names=tab_header)\n",
    "print(len(df_tab))\n",
    "df_tab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54166\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>gencode</th>\n",
       "      <th>count</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>antisense</td>\n",
       "      <td>2125</td>\n",
       "      <td>2125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>bidirectional_promoter_lncRNA</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>2735</td>\n",
       "      <td>2735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>non_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>lncRNA</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>586</td>\n",
       "      <td>586</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    refseq                        gencode  count  transcript_type\n",
       "32  lncRNA                      antisense   2125             2125\n",
       "33  lncRNA  bidirectional_promoter_lncRNA     43               43\n",
       "34  lncRNA                        lincRNA   2735             2735\n",
       "35  lncRNA                     non_coding      1                1\n",
       "36  lncRNA           processed_transcript    586              586"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7713.0\n",
      "{'telomerase_RNA', 'rRNA', 'vault_RNA', 'C_region', 'V_segment', 'scRNA', 'ncRNA_pseudogene', 'lncRNA', 'C_region_pseudogene', 'snRNA', 'tRNA', 'guide_RNA', 'misc_RNA', 'D_segment', 'transcribed_pseudogene', 'J_segment', 'antisense_RNA', 'Y_RNA', 'miRNA', 'protein_coding', 'pseudogene', 'other', 'RNase_MRP_RNA', 'V_segment_pseudogene', 'snoRNA', 'J_segment_pseudogene'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>gencode</th>\n",
       "      <th>count</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>IG_V_gene</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TEC</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>polymorphic_pseudogene</td>\n",
       "      <td>54.0</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>sense_intronic</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>sense_overlapping</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcribed_unitary_pseudogene</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>37.0</td>\n",
       "      <td>37.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>unitary_pseudogene</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>protein_coding</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>27.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            refseq                             gencode  count  transcript_type\n",
       "54  protein_coding                           IG_V_gene    1.0              1.0\n",
       "55  protein_coding                                 TEC    3.0              3.0\n",
       "59  protein_coding              polymorphic_pseudogene   54.0             54.0\n",
       "60  protein_coding                processed_pseudogene    5.0              5.0\n",
       "63  protein_coding                      sense_intronic    1.0              1.0\n",
       "64  protein_coding                   sense_overlapping    3.0              3.0\n",
       "65  protein_coding    transcribed_processed_pseudogene    1.0              1.0\n",
       "66  protein_coding      transcribed_unitary_pseudogene    8.0              8.0\n",
       "67  protein_coding  transcribed_unprocessed_pseudogene   37.0             37.0\n",
       "68  protein_coding                  unitary_pseudogene    2.0              2.0\n",
       "69  protein_coding              unprocessed_pseudogene   27.0             27.0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first we need to replicate the intersection counts and a table for the gencode vs refseq\n",
    "dfg_tab = df_tab[[\"refseqID\",\"gene_biotype\",\"transcript_type\",\"gene_type\"]].groupby(by=[\"gene_biotype\",\"gene_type\"]).count().reset_index()\n",
    "dfg_tab.columns = [\"refseq\",\"gencode\",\"count\",\"transcript_type\"]\n",
    "\n",
    "# define types of interest\n",
    "include_refseq_either = ['protein_coding','lncRNA']\n",
    "include_gencode_either= ['non_coding','bidirectional_promoter_lncRNA','protein_coding','processed_transcript']\n",
    "additional_gencode_for_lncRNA_refseq = ['lincRNA','antisense']\n",
    "\n",
    "# select by conservative thresholds\n",
    "cons_df=dfg_tab[(dfg_tab[\"refseq\"].isin(include_refseq_either))&(dfg_tab[\"gencode\"].isin(include_gencode_either))|\n",
    "                (dfg_tab[\"refseq\"].isin(include_refseq_either))&(dfg_tab[\"gencode\"].isin(additional_gencode_for_lncRNA_refseq))]\n",
    "print(cons_df[\"count\"].sum())\n",
    "display(cons_df.head()) # this is where the conservatively selected transcripts are\n",
    "\n",
    "# now select the non-conservative thresholds\n",
    "noncons_df = dfg_tab[~dfg_tab.isin(cons_df)].dropna()\n",
    "print(noncons_df[\"count\"].sum())\n",
    "print(set(noncons_df[\"refseq\"]))\n",
    "noncons_df[noncons_df[\"refseq\"]==\"protein_coding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54166\n"
     ]
    }
   ],
   "source": [
    "# now need entries from the original tab that correspond to the identified pairs of types\n",
    "df_tab_final = df_tab[(df_tab[\"gene_biotype\"].isin(include_refseq_either))&(df_tab[\"gene_type\"].isin(include_gencode_either))|\n",
    "       (df_tab[\"gene_biotype\"].isin(include_refseq_either))&(df_tab[\"gene_type\"].isin(additional_gencode_for_lncRNA_refseq))]\n",
    "df_tab_final.head()\n",
    "df_tab_final[df_tab_final[\"transcript_type\"]==\"transcribed_unprocessed_pseudogene\"]\n",
    "print(len(df_tab_final))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'exon', 'CDS'}\n",
      "{'J_gene_segment', 'pseudogene', 'mRNA', 'rRNA', 'C_gene_segment', 'lnc_RNA', 'snRNA', 'snoRNA', 'antisense_RNA', 'V_gene_segment', 'transcript', 'guide_RNA'}\n",
      "number of unique parents before subsetting is: 61879\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Curated Genomic</td>\n",
       "      <td>exon</td>\n",
       "      <td>258511</td>\n",
       "      <td>259036</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=gene-RPL23AP21</td>\n",
       "      <td>gene-RPL23AP21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>490756</td>\n",
       "      <td>494898</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_028322.1</td>\n",
       "      <td>rna-NR_028322.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>494992</td>\n",
       "      <td>495049</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_028322.1</td>\n",
       "      <td>rna-NR_028322.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>495277</td>\n",
       "      <td>495445</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_028322.1</td>\n",
       "      <td>rna-NR_028322.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>827591</td>\n",
       "      <td>827775</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_047526.1</td>\n",
       "      <td>rna-NR_047526.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid           source  type   start     end score strand phase  \\\n",
       "0  chr1  Curated Genomic  exon  258511  259036     .      -     .   \n",
       "1  chr1       BestRefSeq  exon  490756  494898     .      -     .   \n",
       "2  chr1       BestRefSeq  exon  494992  495049     .      -     .   \n",
       "3  chr1       BestRefSeq  exon  495277  495445     .      -     .   \n",
       "4  chr1       BestRefSeq  exon  827591  827775     .      +     .   \n",
       "\n",
       "               attributes           parent  \n",
       "0   Parent=gene-RPL23AP21   gene-RPL23AP21  \n",
       "1  Parent=rna-NR_028322.1  rna-NR_028322.1  \n",
       "2  Parent=rna-NR_028322.1  rna-NR_028322.1  \n",
       "3  Parent=rna-NR_028322.1  rna-NR_028322.1  \n",
       "4  Parent=rna-NR_047526.1  rna-NR_047526.1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of transcripts before subsetting is: 61879\n",
      "number of unique transcripts before subsetting is: 61879\n",
      "number of genes before subsetting is 61879\n",
      "number of unique genes before subsetting is: 20892\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transid</th>\n",
       "      <th>geneid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Curated Genomic</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>258511</td>\n",
       "      <td>259036</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-RPL23AP21;geneID=gene-RPL23AP21;gene_n...</td>\n",
       "      <td>gene-RPL23AP21</td>\n",
       "      <td>gene-RPL23AP21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>490756</td>\n",
       "      <td>495445</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-NR_028322.1;geneID=gene-LOC100132287;ge...</td>\n",
       "      <td>rna-NR_028322.1</td>\n",
       "      <td>gene-LOC100132287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>827591</td>\n",
       "      <td>843604</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-NR_047526.1;geneID=gene-LINC01128;gene_...</td>\n",
       "      <td>rna-NR_047526.1</td>\n",
       "      <td>gene-LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>827591</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-NR_047519.1;geneID=gene-LINC01128;gene_...</td>\n",
       "      <td>rna-NR_047519.1</td>\n",
       "      <td>gene-LINC01128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>lnc_RNA</td>\n",
       "      <td>827591</td>\n",
       "      <td>859446</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-NR_047521.1;geneID=gene-LINC01128;gene_...</td>\n",
       "      <td>rna-NR_047521.1</td>\n",
       "      <td>gene-LINC01128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid           source        type   start     end score strand phase  \\\n",
       "0  chr1  Curated Genomic  pseudogene  258511  259036     .      -     .   \n",
       "1  chr1       BestRefSeq     lnc_RNA  490756  495445     .      -     .   \n",
       "2  chr1       BestRefSeq     lnc_RNA  827591  843604     .      +     .   \n",
       "3  chr1       BestRefSeq     lnc_RNA  827591  859446     .      +     .   \n",
       "4  chr1       BestRefSeq     lnc_RNA  827591  859446     .      +     .   \n",
       "\n",
       "                                          attributes          transid  \\\n",
       "0  ID=gene-RPL23AP21;geneID=gene-RPL23AP21;gene_n...   gene-RPL23AP21   \n",
       "1  ID=rna-NR_028322.1;geneID=gene-LOC100132287;ge...  rna-NR_028322.1   \n",
       "2  ID=rna-NR_047526.1;geneID=gene-LINC01128;gene_...  rna-NR_047526.1   \n",
       "3  ID=rna-NR_047519.1;geneID=gene-LINC01128;gene_...  rna-NR_047519.1   \n",
       "4  ID=rna-NR_047521.1;geneID=gene-LINC01128;gene_...  rna-NR_047521.1   \n",
       "\n",
       "              geneid  \n",
       "0     gene-RPL23AP21  \n",
       "1  gene-LOC100132287  \n",
       "2     gene-LINC01128  \n",
       "3     gene-LINC01128  \n",
       "4     gene-LINC01128  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# now need to create a new GFF\n",
    "\n",
    "# begin by separating into a hierarchy\n",
    "has_parent = set(df[df[\"attributes\"].str.contains(\"Parent=\")][\"type\"])\n",
    "print(has_parent)\n",
    "has_no_parent = set(df[~(df[\"attributes\"].str.contains(\"Parent=\"))][\"type\"])\n",
    "print(has_no_parent)\n",
    "lower_df = df[df[\"type\"].isin(has_parent)].reset_index(drop=True) # contains lower end of the hierarchy\n",
    "lower_df[\"parent\"] = lower_df[\"attributes\"].str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "print(\"number of unique parents before subsetting is: \"+str(len(set(lower_df[\"parent\"]))))\n",
    "display(lower_df.head())\n",
    "upper_df = df[df[\"type\"].isin(has_no_parent)].reset_index(drop=True) #contains upper end of the hierarchy\n",
    "upper_df[\"transid\"] = upper_df[\"attributes\"].str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "upper_df[\"geneid\"] = upper_df[\"attributes\"].str.split(\"geneID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "print(\"number of transcripts before subsetting is: \"+str(len(upper_df[\"transid\"])))\n",
    "print(\"number of unique transcripts before subsetting is: \"+str(len(set(upper_df[\"transid\"]))))\n",
    "print(\"number of genes before subsetting is \"+str(len(upper_df[\"geneid\"])))\n",
    "print(\"number of unique genes before subsetting is: \"+str(len(set(upper_df[\"geneid\"]))))\n",
    "display(upper_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54166\n",
      "982115\n",
      "54166\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>827591</td>\n",
       "      <td>827775</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_047526.1</td>\n",
       "      <td>rna-NR_047526.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>829003</td>\n",
       "      <td>829104</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_047526.1</td>\n",
       "      <td>rna-NR_047526.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>841200</td>\n",
       "      <td>843604</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_047526.1</td>\n",
       "      <td>rna-NR_047526.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>827591</td>\n",
       "      <td>827775</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_047519.1</td>\n",
       "      <td>rna-NR_047519.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>829003</td>\n",
       "      <td>829104</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-NR_047519.1</td>\n",
       "      <td>rna-NR_047519.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid      source  type   start     end score strand phase  \\\n",
       "4  chr1  BestRefSeq  exon  827591  827775     .      +     .   \n",
       "5  chr1  BestRefSeq  exon  829003  829104     .      +     .   \n",
       "6  chr1  BestRefSeq  exon  841200  843604     .      +     .   \n",
       "7  chr1  BestRefSeq  exon  827591  827775     .      +     .   \n",
       "8  chr1  BestRefSeq  exon  829003  829104     .      +     .   \n",
       "\n",
       "               attributes           parent  \n",
       "4  Parent=rna-NR_047526.1  rna-NR_047526.1  \n",
       "5  Parent=rna-NR_047526.1  rna-NR_047526.1  \n",
       "6  Parent=rna-NR_047526.1  rna-NR_047526.1  \n",
       "7  Parent=rna-NR_047519.1  rna-NR_047519.1  \n",
       "8  Parent=rna-NR_047519.1  rna-NR_047519.1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to get the correct genes\n",
    "\n",
    "# first select transcripts\n",
    "upper_df = upper_df[(upper_df[\"transid\"].isin(set(df_tab_final[\"gencodeID\"])))|\n",
    "                    (upper_df[\"transid\"].isin(set(df_tab_final[\"refseqID\"])))]\n",
    "upper_df[\"type\"] = \"transcript\"\n",
    "print(len(upper_df))\n",
    "upper_df.head()\n",
    "\n",
    "# second - select exons\n",
    "lower_df = lower_df[(lower_df[\"parent\"].isin(set(df_tab_final[\"gencodeID\"])))|\n",
    "                    (lower_df[\"parent\"].isin(set(df_tab_final[\"refseqID\"])))]\n",
    "print(len(lower_df))\n",
    "print(len(set(lower_df[\"parent\"])))\n",
    "lower_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sparrow/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "# now to join the upper and lower\n",
    "full_df = pd.concat([upper_df,lower_df],axis=0)\n",
    "# now to get the parent and transid in the same column\n",
    "full_df[\"parent\"] = np.where(full_df[\"parent\"].isnull(),full_df[\"transid\"],full_df[\"parent\"])\n",
    "# now to sort the transcripts and exons and CDS accordingly\n",
    "full_df[\"type\"]=pd.Categorical(full_df[\"type\"],categories=[\"transcript\",\"exon\",\"CDS\"],ordered=True)\n",
    "full_df=full_df.sort_values(by=[\"parent\",\"type\"]).reset_index(drop=True)\n",
    "# now to select only the relevant columns and order them\n",
    "full_df = full_df[gff3Cols]\n",
    "full_df.to_csv(\"./prepAnnotation/refseq_vs_gencode_intersection.prot_lnc.gff\",sep='\\t',index=False,header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>127588345</td>\n",
       "      <td>127591705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000233.9;gene_id=ENSG00000004059.1...</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8940365</td>\n",
       "      <td>8949955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000412.7;gene_id=ENSG00000003056.7...</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72148038</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000001146.6;gene_id=ENSG00000003137.8...</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11393150</td>\n",
       "      <td>11429765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002596.5;gene_id=ENSG00000002587.9...</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50155129</td>\n",
       "      <td>50189075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002829.7;gene_id=ENSG00000001617.1...</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid  source        type      start        end score strand phase  \\\n",
       "0   chr7  HAVANA  transcript  127588345  127591705     .      +     .   \n",
       "1  chr12  HAVANA  transcript    8940365    8949955     .      -     .   \n",
       "2   chr2  HAVANA  transcript   72129238   72148038     .      -     .   \n",
       "3   chr4  HAVANA  transcript   11393150   11429765     .      -     .   \n",
       "4   chr3  HAVANA  transcript   50155129   50189075     .      +     .   \n",
       "\n",
       "                                          attributes             gene_id  \n",
       "0  ID=ENST00000000233.9;gene_id=ENSG00000004059.1...  ENSG00000004059.10  \n",
       "1  ID=ENST00000000412.7;gene_id=ENSG00000003056.7...   ENSG00000003056.7  \n",
       "2  ID=ENST00000001146.6;gene_id=ENSG00000003137.8...   ENSG00000003137.8  \n",
       "3  ID=ENST00000002596.5;gene_id=ENSG00000002587.9...   ENSG00000002587.9  \n",
       "4  ID=ENST00000002829.7;gene_id=ENSG00000001617.1...  ENSG00000001617.11  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now need to add genes to the intersection\n",
    "full_df_tr = full_df[full_df[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "full_df_tr[\"gene_id\"] = full_df_tr.attributes.str.split(\"gene_id=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "full_df_tr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'HAVANA', 'ENSEMBL'}\n",
      "{'Gnomon', 'BestRefSeq', 'Curated Genomic'}\n"
     ]
    }
   ],
   "source": [
    "set_geneID = set(full_df_tr[~full_df_tr[\"gene_id\"].isnull()][\"source\"])\n",
    "print(set_geneID)\n",
    "set_no_geneID = set(full_df_tr[full_df_tr[\"gene_id\"].isnull()][\"source\"])\n",
    "print(set_no_geneID)\n",
    "assert len(set_geneID.intersection(set_no_geneID)) == 0, \"there's an overlap\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>geneID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>127588345</td>\n",
       "      <td>127591705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000233.9;gene_id=ENSG00000004059.1...</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8940365</td>\n",
       "      <td>8949955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000412.7;gene_id=ENSG00000003056.7...</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72148038</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000001146.6;gene_id=ENSG00000003137.8...</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11393150</td>\n",
       "      <td>11429765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002596.5;gene_id=ENSG00000002587.9...</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50155129</td>\n",
       "      <td>50189075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002829.7;gene_id=ENSG00000001617.1...</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid  source        type      start        end score strand phase  \\\n",
       "0   chr7  HAVANA  transcript  127588345  127591705     .      +     .   \n",
       "1  chr12  HAVANA  transcript    8940365    8949955     .      -     .   \n",
       "2   chr2  HAVANA  transcript   72129238   72148038     .      -     .   \n",
       "3   chr4  HAVANA  transcript   11393150   11429765     .      -     .   \n",
       "4   chr3  HAVANA  transcript   50155129   50189075     .      +     .   \n",
       "\n",
       "                                          attributes             gene_id  \\\n",
       "0  ID=ENST00000000233.9;gene_id=ENSG00000004059.1...  ENSG00000004059.10   \n",
       "1  ID=ENST00000000412.7;gene_id=ENSG00000003056.7...   ENSG00000003056.7   \n",
       "2  ID=ENST00000001146.6;gene_id=ENSG00000003137.8...   ENSG00000003137.8   \n",
       "3  ID=ENST00000002596.5;gene_id=ENSG00000002587.9...   ENSG00000002587.9   \n",
       "4  ID=ENST00000002829.7;gene_id=ENSG00000001617.1...  ENSG00000001617.11   \n",
       "\n",
       "               geneID  \n",
       "0  ENSG00000004059.10  \n",
       "1   ENSG00000003056.7  \n",
       "2   ENSG00000003137.8  \n",
       "3   ENSG00000002587.9  \n",
       "4  ENSG00000001617.11  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_df_tr[\"geneID\"] = full_df_tr.attributes.str.split(\"geneID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "full_df_tr[\"geneID\"] = np.where(full_df_tr.geneID.isnull(),\n",
    "                        full_df_tr.attributes.str.split(\"gene_id=\",expand=True)[1].str.split(\";\",expand=True)[0],\n",
    "                        full_df_tr.geneID)\n",
    "full_df_tr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(full_df_tr[full_df_tr[\"geneID\"].isnull()]) == 0, \"some gene ids have not been extracted\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>geneID</th>\n",
       "      <th>trans_parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>127588345</td>\n",
       "      <td>127591705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000233.9;gene_id=ENSG00000004059.1...</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8940365</td>\n",
       "      <td>8949955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000412.7;gene_id=ENSG00000003056.7...</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72148038</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000001146.6;gene_id=ENSG00000003137.8...</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11393150</td>\n",
       "      <td>11429765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002596.5;gene_id=ENSG00000002587.9...</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50155129</td>\n",
       "      <td>50189075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002829.7;gene_id=ENSG00000001617.1...</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid  source        type      start        end score strand phase  \\\n",
       "0   chr7  HAVANA  transcript  127588345  127591705     .      +     .   \n",
       "1  chr12  HAVANA  transcript    8940365    8949955     .      -     .   \n",
       "2   chr2  HAVANA  transcript   72129238   72148038     .      -     .   \n",
       "3   chr4  HAVANA  transcript   11393150   11429765     .      -     .   \n",
       "4   chr3  HAVANA  transcript   50155129   50189075     .      +     .   \n",
       "\n",
       "                                          attributes             gene_id  \\\n",
       "0  ID=ENST00000000233.9;gene_id=ENSG00000004059.1...  ENSG00000004059.10   \n",
       "1  ID=ENST00000000412.7;gene_id=ENSG00000003056.7...   ENSG00000003056.7   \n",
       "2  ID=ENST00000001146.6;gene_id=ENSG00000003137.8...   ENSG00000003137.8   \n",
       "3  ID=ENST00000002596.5;gene_id=ENSG00000002587.9...   ENSG00000002587.9   \n",
       "4  ID=ENST00000002829.7;gene_id=ENSG00000001617.1...  ENSG00000001617.11   \n",
       "\n",
       "               geneID        trans_parent  \n",
       "0  ENSG00000004059.10  ENSG00000004059.10  \n",
       "1   ENSG00000003056.7   ENSG00000003056.7  \n",
       "2   ENSG00000003137.8   ENSG00000003137.8  \n",
       "3   ENSG00000002587.9   ENSG00000002587.9  \n",
       "4  ENSG00000001617.11  ENSG00000001617.11  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now we need merge two types of geneIDs together\n",
    "full_df_tr[\"trans_parent\"] = full_df_tr.geneID\n",
    "full_df_tr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of genes = 32522\n",
      "number of genes in upper: 17282\n",
      "number of geneID extracted genes from full_df_tr: 32522\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "17282"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"number of genes = \"+str(len(set(full_df_tr[\"trans_parent\"]))))\n",
    "print(\"number of genes in upper: \"+str(len(set(upper_df[\"geneid\"]))))\n",
    "print(\"number of geneID extracted genes from full_df_tr: \"+str(len(set(full_df_tr[\"geneID\"]))))\n",
    "len(set(full_df_tr[\"attributes\"].str.split(\"geneID=\",expand=True)[1].str.split(\";\",expand=True)[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169893959</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000457.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>169794730</td>\n",
       "      <td>169854080</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000460.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>27612064</td>\n",
       "      <td>27635277</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000938.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>196651878</td>\n",
       "      <td>196747504</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000971.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>53497341</td>\n",
       "      <td>53545129</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000001084.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source  type      start        end score strand phase  \\\n",
       "0  chr1  intersect  gene  169849631  169893959     .      -     .   \n",
       "1  chr1  intersect  gene  169794730  169854080     .      +     .   \n",
       "2  chr1  intersect  gene   27612064   27635277     .      -     .   \n",
       "3  chr1  intersect  gene  196651878  196747504     .      +     .   \n",
       "4  chr6  intersect  gene   53497341   53545129     .      -     .   \n",
       "\n",
       "              attributes  \n",
       "0  ID=ENSG00000000457.13  \n",
       "1  ID=ENSG00000000460.16  \n",
       "2  ID=ENSG00000000938.12  \n",
       "3  ID=ENSG00000000971.15  \n",
       "4  ID=ENSG00000001084.12  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now let's see if we can merge transcripts together by genes\n",
    "gene_df = full_df_tr.groupby(by=\"trans_parent\").agg({\"start\":\"min\",\n",
    "                                                   \"end\":\"max\",\n",
    "                                                   \"strand\":lambda row: \",\".join(list(set(row))),\n",
    "                                                   \"seqid\":lambda row: \",\".join(list(set(row))),\n",
    "                                                   \"geneID\":\"count\"}).reset_index()\n",
    "gene_df[\"score\"] = \".\"\n",
    "gene_df[\"phase\"] = \".\"\n",
    "gene_df[\"source\"]= \"intersect\"\n",
    "gene_df[\"type\"] = \"gene\"\n",
    "gene_df[\"attributes\"] = \"ID=\"+gene_df[\"trans_parent\"]\n",
    "gene_df = gene_df[gff3Cols]\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>geneID</th>\n",
       "      <th>trans_parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>127588345</td>\n",
       "      <td>127591705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000233.9;Parent=ENSG00000004059.10...</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8940365</td>\n",
       "      <td>8949955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000412.7;Parent=ENSG00000003056.7;...</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72148038</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000001146.6;Parent=ENSG00000003137.8;...</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11393150</td>\n",
       "      <td>11429765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002596.5;Parent=ENSG00000002587.9;...</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50155129</td>\n",
       "      <td>50189075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002829.7;Parent=ENSG00000001617.11...</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid  source        type      start        end score strand phase  \\\n",
       "0   chr7  HAVANA  transcript  127588345  127591705     .      +     .   \n",
       "1  chr12  HAVANA  transcript    8940365    8949955     .      -     .   \n",
       "2   chr2  HAVANA  transcript   72129238   72148038     .      -     .   \n",
       "3   chr4  HAVANA  transcript   11393150   11429765     .      -     .   \n",
       "4   chr3  HAVANA  transcript   50155129   50189075     .      +     .   \n",
       "\n",
       "                                          attributes             gene_id  \\\n",
       "0  ID=ENST00000000233.9;Parent=ENSG00000004059.10...  ENSG00000004059.10   \n",
       "1  ID=ENST00000000412.7;Parent=ENSG00000003056.7;...   ENSG00000003056.7   \n",
       "2  ID=ENST00000001146.6;Parent=ENSG00000003137.8;...   ENSG00000003137.8   \n",
       "3  ID=ENST00000002596.5;Parent=ENSG00000002587.9;...   ENSG00000002587.9   \n",
       "4  ID=ENST00000002829.7;Parent=ENSG00000001617.11...  ENSG00000001617.11   \n",
       "\n",
       "               geneID        trans_parent  \n",
       "0  ENSG00000004059.10  ENSG00000004059.10  \n",
       "1   ENSG00000003056.7   ENSG00000003056.7  \n",
       "2   ENSG00000003137.8   ENSG00000003137.8  \n",
       "3   ENSG00000002587.9   ENSG00000002587.9  \n",
       "4  ENSG00000001617.11  ENSG00000001617.11  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now to modify the transcript entries to include Parent attribute\n",
    "full_df_tr[\"attributes\"] = full_df_tr.attributes.str.split(\";\",expand=True,n=1)[0]+\\\n",
    "                            \";Parent=\"+full_df_tr[\"trans_parent\"]+\";\"+\\\n",
    "                            full_df_tr.attributes.str.split(\";\",expand=True,n=1)[1]\n",
    "full_df_tr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169893959</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000457.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>169794730</td>\n",
       "      <td>169854080</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000460.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>27612064</td>\n",
       "      <td>27635277</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000938.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>196651878</td>\n",
       "      <td>196747504</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000971.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>53497341</td>\n",
       "      <td>53545129</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000001084.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source  type      start        end score strand phase  \\\n",
       "0  chr1  intersect  gene  169849631  169893959     .      -     .   \n",
       "1  chr1  intersect  gene  169794730  169854080     .      +     .   \n",
       "2  chr1  intersect  gene   27612064   27635277     .      -     .   \n",
       "3  chr1  intersect  gene  196651878  196747504     .      +     .   \n",
       "4  chr6  intersect  gene   53497341   53545129     .      -     .   \n",
       "\n",
       "              attributes  \n",
       "0  ID=ENSG00000000457.13  \n",
       "1  ID=ENSG00000000460.16  \n",
       "2  ID=ENSG00000000938.12  \n",
       "3  ID=ENSG00000000971.15  \n",
       "4  ID=ENSG00000001084.12  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now separate out exons and CDS and concatenate them with genes and transcripts and then sort accordingly\n",
    "lower_df = full_df[full_df[\"type\"].isin(has_parent)].reset_index(drop=True)\n",
    "lower_df = lower_df[gff3Cols]\n",
    "full_df_tr = full_df_tr[gff3Cols]\n",
    "resdf = pd.concat([gene_df,full_df_tr,lower_df],axis=0)\n",
    "resdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>127588345</td>\n",
       "      <td>127588565</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>127589083</td>\n",
       "      <td>127589163</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>127589485</td>\n",
       "      <td>127589594</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>127590066</td>\n",
       "      <td>127590137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>127590963</td>\n",
       "      <td>127591088</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>127591213</td>\n",
       "      <td>127591705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127588499</td>\n",
       "      <td>127588565</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127589083</td>\n",
       "      <td>127589163</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127589485</td>\n",
       "      <td>127589594</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127590066</td>\n",
       "      <td>127590137</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127590963</td>\n",
       "      <td>127591088</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127591213</td>\n",
       "      <td>127591299</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=ENST00000000233.9</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>8940365</td>\n",
       "      <td>8941940</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>8942416</td>\n",
       "      <td>8942542</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>8943405</td>\n",
       "      <td>8943535</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>8943801</td>\n",
       "      <td>8943910</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>8945418</td>\n",
       "      <td>8945584</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>8946229</td>\n",
       "      <td>8946405</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>8949488</td>\n",
       "      <td>8949955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8941818</td>\n",
       "      <td>8941940</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8942416</td>\n",
       "      <td>8942542</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8943405</td>\n",
       "      <td>8943535</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8943801</td>\n",
       "      <td>8943910</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8945418</td>\n",
       "      <td>8945584</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8946229</td>\n",
       "      <td>8946404</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>Parent=ENST00000000412.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72132619</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000001146.6</td>\n",
       "      <td>ENST00000001146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>72133023</td>\n",
       "      <td>72133307</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000001146.6</td>\n",
       "      <td>ENST00000001146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>72134761</td>\n",
       "      <td>72134916</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000001146.6</td>\n",
       "      <td>ENST00000001146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>72135144</td>\n",
       "      <td>72135419</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000001146.6</td>\n",
       "      <td>ENST00000001146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>72143989</td>\n",
       "      <td>72144213</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=ENST00000001146.6</td>\n",
       "      <td>ENST00000001146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982085</th>\n",
       "      <td>chr19_GL949746v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>195328</td>\n",
       "      <td>195630</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952182.3</td>\n",
       "      <td>rna-XR_952182.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982086</th>\n",
       "      <td>chr19_GL949746v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>195883</td>\n",
       "      <td>196179</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952182.3</td>\n",
       "      <td>rna-XR_952182.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982087</th>\n",
       "      <td>chr19_GL949746v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>196628</td>\n",
       "      <td>196930</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952182.3</td>\n",
       "      <td>rna-XR_952182.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982088</th>\n",
       "      <td>chr19_GL949746v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>197078</td>\n",
       "      <td>197362</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952182.3</td>\n",
       "      <td>rna-XR_952182.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982089</th>\n",
       "      <td>chr19_GL949746v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>197547</td>\n",
       "      <td>197582</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952182.3</td>\n",
       "      <td>rna-XR_952182.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982090</th>\n",
       "      <td>chr19_GL949746v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>197743</td>\n",
       "      <td>197856</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952182.3</td>\n",
       "      <td>rna-XR_952182.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982091</th>\n",
       "      <td>chr6_GL000251v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>983927</td>\n",
       "      <td>983991</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952235.2</td>\n",
       "      <td>rna-XR_952235.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982092</th>\n",
       "      <td>chr6_GL000251v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>991773</td>\n",
       "      <td>991829</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952235.2</td>\n",
       "      <td>rna-XR_952235.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982093</th>\n",
       "      <td>chr6_GL000251v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>993621</td>\n",
       "      <td>994120</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952235.2</td>\n",
       "      <td>rna-XR_952235.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982094</th>\n",
       "      <td>chr6_GL000251v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>996720</td>\n",
       "      <td>997010</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952235.2</td>\n",
       "      <td>rna-XR_952235.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982095</th>\n",
       "      <td>chr8_KI270901v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>106445</td>\n",
       "      <td>106681</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952326.2</td>\n",
       "      <td>rna-XR_952326.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982096</th>\n",
       "      <td>chr8_KI270901v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>115026</td>\n",
       "      <td>115058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952326.2</td>\n",
       "      <td>rna-XR_952326.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982097</th>\n",
       "      <td>chr8_KI270901v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>115844</td>\n",
       "      <td>115983</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952326.2</td>\n",
       "      <td>rna-XR_952326.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982098</th>\n",
       "      <td>chr8_KI270901v1_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>118590</td>\n",
       "      <td>118755</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952326.2</td>\n",
       "      <td>rna-XR_952326.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982099</th>\n",
       "      <td>chr6_GL000252v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>762998</td>\n",
       "      <td>763062</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952695.3</td>\n",
       "      <td>rna-XR_952695.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982100</th>\n",
       "      <td>chr6_GL000252v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>770844</td>\n",
       "      <td>770900</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952695.3</td>\n",
       "      <td>rna-XR_952695.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982101</th>\n",
       "      <td>chr6_GL000252v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>772692</td>\n",
       "      <td>773191</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952695.3</td>\n",
       "      <td>rna-XR_952695.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982102</th>\n",
       "      <td>chr6_GL000252v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>775791</td>\n",
       "      <td>776079</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952695.3</td>\n",
       "      <td>rna-XR_952695.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982103</th>\n",
       "      <td>chr6_GL000253v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>762601</td>\n",
       "      <td>762665</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952881.3</td>\n",
       "      <td>rna-XR_952881.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982104</th>\n",
       "      <td>chr6_GL000253v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>770446</td>\n",
       "      <td>770502</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952881.3</td>\n",
       "      <td>rna-XR_952881.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982105</th>\n",
       "      <td>chr6_GL000253v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>772294</td>\n",
       "      <td>772793</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952881.3</td>\n",
       "      <td>rna-XR_952881.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982106</th>\n",
       "      <td>chr6_GL000253v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>775393</td>\n",
       "      <td>775681</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952881.3</td>\n",
       "      <td>rna-XR_952881.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982107</th>\n",
       "      <td>chr6_GL000254v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>762850</td>\n",
       "      <td>762914</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952962.3</td>\n",
       "      <td>rna-XR_952962.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982108</th>\n",
       "      <td>chr6_GL000254v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>770696</td>\n",
       "      <td>770752</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952962.3</td>\n",
       "      <td>rna-XR_952962.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982109</th>\n",
       "      <td>chr6_GL000254v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>772544</td>\n",
       "      <td>773043</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952962.3</td>\n",
       "      <td>rna-XR_952962.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982110</th>\n",
       "      <td>chr6_GL000254v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>775643</td>\n",
       "      <td>775931</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_952962.3</td>\n",
       "      <td>rna-XR_952962.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982111</th>\n",
       "      <td>chr6_GL000255v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>762942</td>\n",
       "      <td>763006</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_953031.3</td>\n",
       "      <td>rna-XR_953031.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982112</th>\n",
       "      <td>chr6_GL000255v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>770787</td>\n",
       "      <td>770843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_953031.3</td>\n",
       "      <td>rna-XR_953031.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982113</th>\n",
       "      <td>chr6_GL000255v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>772635</td>\n",
       "      <td>773134</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_953031.3</td>\n",
       "      <td>rna-XR_953031.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982114</th>\n",
       "      <td>chr6_GL000255v2_alt</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>775734</td>\n",
       "      <td>776022</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=rna-XR_953031.3</td>\n",
       "      <td>rna-XR_953031.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>982115 rows  10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       seqid  source  type      start        end score strand  \\\n",
       "0                       chr7  HAVANA  exon  127588345  127588565     .      +   \n",
       "1                       chr7  HAVANA  exon  127589083  127589163     .      +   \n",
       "2                       chr7  HAVANA  exon  127589485  127589594     .      +   \n",
       "3                       chr7  HAVANA  exon  127590066  127590137     .      +   \n",
       "4                       chr7  HAVANA  exon  127590963  127591088     .      +   \n",
       "5                       chr7  HAVANA  exon  127591213  127591705     .      +   \n",
       "6                       chr7  HAVANA   CDS  127588499  127588565     .      +   \n",
       "7                       chr7  HAVANA   CDS  127589083  127589163     .      +   \n",
       "8                       chr7  HAVANA   CDS  127589485  127589594     .      +   \n",
       "9                       chr7  HAVANA   CDS  127590066  127590137     .      +   \n",
       "10                      chr7  HAVANA   CDS  127590963  127591088     .      +   \n",
       "11                      chr7  HAVANA   CDS  127591213  127591299     .      +   \n",
       "12                     chr12  HAVANA  exon    8940365    8941940     .      -   \n",
       "13                     chr12  HAVANA  exon    8942416    8942542     .      -   \n",
       "14                     chr12  HAVANA  exon    8943405    8943535     .      -   \n",
       "15                     chr12  HAVANA  exon    8943801    8943910     .      -   \n",
       "16                     chr12  HAVANA  exon    8945418    8945584     .      -   \n",
       "17                     chr12  HAVANA  exon    8946229    8946405     .      -   \n",
       "18                     chr12  HAVANA  exon    8949488    8949955     .      -   \n",
       "19                     chr12  HAVANA   CDS    8941818    8941940     .      -   \n",
       "20                     chr12  HAVANA   CDS    8942416    8942542     .      -   \n",
       "21                     chr12  HAVANA   CDS    8943405    8943535     .      -   \n",
       "22                     chr12  HAVANA   CDS    8943801    8943910     .      -   \n",
       "23                     chr12  HAVANA   CDS    8945418    8945584     .      -   \n",
       "24                     chr12  HAVANA   CDS    8946229    8946404     .      -   \n",
       "25                      chr2  HAVANA  exon   72129238   72132619     .      -   \n",
       "26                      chr2  HAVANA  exon   72133023   72133307     .      -   \n",
       "27                      chr2  HAVANA  exon   72134761   72134916     .      -   \n",
       "28                      chr2  HAVANA  exon   72135144   72135419     .      -   \n",
       "29                      chr2  HAVANA  exon   72143989   72144213     .      -   \n",
       "...                      ...     ...   ...        ...        ...   ...    ...   \n",
       "982085  chr19_GL949746v1_alt  Gnomon  exon     195328     195630     .      -   \n",
       "982086  chr19_GL949746v1_alt  Gnomon  exon     195883     196179     .      -   \n",
       "982087  chr19_GL949746v1_alt  Gnomon  exon     196628     196930     .      -   \n",
       "982088  chr19_GL949746v1_alt  Gnomon  exon     197078     197362     .      -   \n",
       "982089  chr19_GL949746v1_alt  Gnomon  exon     197547     197582     .      -   \n",
       "982090  chr19_GL949746v1_alt  Gnomon  exon     197743     197856     .      -   \n",
       "982091   chr6_GL000251v2_alt  Gnomon  exon     983927     983991     .      +   \n",
       "982092   chr6_GL000251v2_alt  Gnomon  exon     991773     991829     .      +   \n",
       "982093   chr6_GL000251v2_alt  Gnomon  exon     993621     994120     .      +   \n",
       "982094   chr6_GL000251v2_alt  Gnomon  exon     996720     997010     .      +   \n",
       "982095   chr8_KI270901v1_alt  Gnomon  exon     106445     106681     .      +   \n",
       "982096   chr8_KI270901v1_alt  Gnomon  exon     115026     115058     .      +   \n",
       "982097   chr8_KI270901v1_alt  Gnomon  exon     115844     115983     .      +   \n",
       "982098   chr8_KI270901v1_alt  Gnomon  exon     118590     118755     .      +   \n",
       "982099   chr6_GL000252v2_alt  Gnomon  exon     762998     763062     .      +   \n",
       "982100   chr6_GL000252v2_alt  Gnomon  exon     770844     770900     .      +   \n",
       "982101   chr6_GL000252v2_alt  Gnomon  exon     772692     773191     .      +   \n",
       "982102   chr6_GL000252v2_alt  Gnomon  exon     775791     776079     .      +   \n",
       "982103   chr6_GL000253v2_alt  Gnomon  exon     762601     762665     .      +   \n",
       "982104   chr6_GL000253v2_alt  Gnomon  exon     770446     770502     .      +   \n",
       "982105   chr6_GL000253v2_alt  Gnomon  exon     772294     772793     .      +   \n",
       "982106   chr6_GL000253v2_alt  Gnomon  exon     775393     775681     .      +   \n",
       "982107   chr6_GL000254v2_alt  Gnomon  exon     762850     762914     .      +   \n",
       "982108   chr6_GL000254v2_alt  Gnomon  exon     770696     770752     .      +   \n",
       "982109   chr6_GL000254v2_alt  Gnomon  exon     772544     773043     .      +   \n",
       "982110   chr6_GL000254v2_alt  Gnomon  exon     775643     775931     .      +   \n",
       "982111   chr6_GL000255v2_alt  Gnomon  exon     762942     763006     .      +   \n",
       "982112   chr6_GL000255v2_alt  Gnomon  exon     770787     770843     .      +   \n",
       "982113   chr6_GL000255v2_alt  Gnomon  exon     772635     773134     .      +   \n",
       "982114   chr6_GL000255v2_alt  Gnomon  exon     775734     776022     .      +   \n",
       "\n",
       "       phase                attributes            transID  \n",
       "0          .  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "1          .  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "2          .  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "3          .  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "4          .  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "5          .  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "6          0  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "7          2  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "8          2  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "9          0  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "10         0  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "11         0  Parent=ENST00000000233.9  ENST00000000233.9  \n",
       "12         .  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "13         .  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "14         .  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "15         .  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "16         .  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "17         .  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "18         .  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "19         0  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "20         1  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "21         0  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "22         2  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "23         1  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "24         0  Parent=ENST00000000412.7  ENST00000000412.7  \n",
       "25         .  Parent=ENST00000001146.6  ENST00000001146.6  \n",
       "26         .  Parent=ENST00000001146.6  ENST00000001146.6  \n",
       "27         .  Parent=ENST00000001146.6  ENST00000001146.6  \n",
       "28         .  Parent=ENST00000001146.6  ENST00000001146.6  \n",
       "29         .  Parent=ENST00000001146.6  ENST00000001146.6  \n",
       "...      ...                       ...                ...  \n",
       "982085     .    Parent=rna-XR_952182.3    rna-XR_952182.3  \n",
       "982086     .    Parent=rna-XR_952182.3    rna-XR_952182.3  \n",
       "982087     .    Parent=rna-XR_952182.3    rna-XR_952182.3  \n",
       "982088     .    Parent=rna-XR_952182.3    rna-XR_952182.3  \n",
       "982089     .    Parent=rna-XR_952182.3    rna-XR_952182.3  \n",
       "982090     .    Parent=rna-XR_952182.3    rna-XR_952182.3  \n",
       "982091     .    Parent=rna-XR_952235.2    rna-XR_952235.2  \n",
       "982092     .    Parent=rna-XR_952235.2    rna-XR_952235.2  \n",
       "982093     .    Parent=rna-XR_952235.2    rna-XR_952235.2  \n",
       "982094     .    Parent=rna-XR_952235.2    rna-XR_952235.2  \n",
       "982095     .    Parent=rna-XR_952326.2    rna-XR_952326.2  \n",
       "982096     .    Parent=rna-XR_952326.2    rna-XR_952326.2  \n",
       "982097     .    Parent=rna-XR_952326.2    rna-XR_952326.2  \n",
       "982098     .    Parent=rna-XR_952326.2    rna-XR_952326.2  \n",
       "982099     .    Parent=rna-XR_952695.3    rna-XR_952695.3  \n",
       "982100     .    Parent=rna-XR_952695.3    rna-XR_952695.3  \n",
       "982101     .    Parent=rna-XR_952695.3    rna-XR_952695.3  \n",
       "982102     .    Parent=rna-XR_952695.3    rna-XR_952695.3  \n",
       "982103     .    Parent=rna-XR_952881.3    rna-XR_952881.3  \n",
       "982104     .    Parent=rna-XR_952881.3    rna-XR_952881.3  \n",
       "982105     .    Parent=rna-XR_952881.3    rna-XR_952881.3  \n",
       "982106     .    Parent=rna-XR_952881.3    rna-XR_952881.3  \n",
       "982107     .    Parent=rna-XR_952962.3    rna-XR_952962.3  \n",
       "982108     .    Parent=rna-XR_952962.3    rna-XR_952962.3  \n",
       "982109     .    Parent=rna-XR_952962.3    rna-XR_952962.3  \n",
       "982110     .    Parent=rna-XR_952962.3    rna-XR_952962.3  \n",
       "982111     .    Parent=rna-XR_953031.3    rna-XR_953031.3  \n",
       "982112     .    Parent=rna-XR_953031.3    rna-XR_953031.3  \n",
       "982113     .    Parent=rna-XR_953031.3    rna-XR_953031.3  \n",
       "982114     .    Parent=rna-XR_953031.3    rna-XR_953031.3  \n",
       "\n",
       "[982115 rows x 10 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lastly we need to sort the gff correctly\n",
    "\n",
    "lower_df[\"transID\"]=lower_df.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "assert len(lower_df[lower_df[\"transID\"].isnull()])==0,\"some parents are missing from exons\"\n",
    "lower_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>geneID</th>\n",
       "      <th>transID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>127588345</td>\n",
       "      <td>127591705</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000233.9;Parent=ENSG00000004059.10...</td>\n",
       "      <td>ENSG00000004059.10</td>\n",
       "      <td>ENST00000000233.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8940365</td>\n",
       "      <td>8949955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000000412.7;Parent=ENSG00000003056.7;...</td>\n",
       "      <td>ENSG00000003056.7</td>\n",
       "      <td>ENST00000000412.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72129238</td>\n",
       "      <td>72148038</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000001146.6;Parent=ENSG00000003137.8;...</td>\n",
       "      <td>ENSG00000003137.8</td>\n",
       "      <td>ENST00000001146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11393150</td>\n",
       "      <td>11429765</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002596.5;Parent=ENSG00000002587.9;...</td>\n",
       "      <td>ENSG00000002587.9</td>\n",
       "      <td>ENST00000002596.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>50155129</td>\n",
       "      <td>50189075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000002829.7;Parent=ENSG00000001617.11...</td>\n",
       "      <td>ENSG00000001617.11</td>\n",
       "      <td>ENST00000002829.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid  source        type      start        end score strand phase  \\\n",
       "0   chr7  HAVANA  transcript  127588345  127591705     .      +     .   \n",
       "1  chr12  HAVANA  transcript    8940365    8949955     .      -     .   \n",
       "2   chr2  HAVANA  transcript   72129238   72148038     .      -     .   \n",
       "3   chr4  HAVANA  transcript   11393150   11429765     .      -     .   \n",
       "4   chr3  HAVANA  transcript   50155129   50189075     .      +     .   \n",
       "\n",
       "                                          attributes              geneID  \\\n",
       "0  ID=ENST00000000233.9;Parent=ENSG00000004059.10...  ENSG00000004059.10   \n",
       "1  ID=ENST00000000412.7;Parent=ENSG00000003056.7;...   ENSG00000003056.7   \n",
       "2  ID=ENST00000001146.6;Parent=ENSG00000003137.8;...   ENSG00000003137.8   \n",
       "3  ID=ENST00000002596.5;Parent=ENSG00000002587.9;...   ENSG00000002587.9   \n",
       "4  ID=ENST00000002829.7;Parent=ENSG00000001617.11...  ENSG00000001617.11   \n",
       "\n",
       "             transID  \n",
       "0  ENST00000000233.9  \n",
       "1  ENST00000000412.7  \n",
       "2  ENST00000001146.6  \n",
       "3  ENST00000002596.5  \n",
       "4  ENST00000002829.7  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_df_tr[\"geneID\"]=full_df_tr.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "assert len(full_df_tr[full_df_tr[\"geneID\"].isnull()])==0,\"some parents are missing from transcripts\"\n",
    "full_df_tr[\"transID\"]=full_df_tr.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "assert len(full_df_tr[full_df_tr[\"transID\"].isnull()])==0,\"some IDs are missing from transcripts\"\n",
    "full_df_tr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>geneID</th>\n",
       "      <th>transID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>169849631</td>\n",
       "      <td>169893959</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000457.13</td>\n",
       "      <td>ENSG00000000457.13</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>169794730</td>\n",
       "      <td>169854080</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000460.16</td>\n",
       "      <td>ENSG00000000460.16</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>27612064</td>\n",
       "      <td>27635277</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000938.12</td>\n",
       "      <td>ENSG00000000938.12</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>196651878</td>\n",
       "      <td>196747504</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000000971.15</td>\n",
       "      <td>ENSG00000000971.15</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6</td>\n",
       "      <td>intersect</td>\n",
       "      <td>gene</td>\n",
       "      <td>53497341</td>\n",
       "      <td>53545129</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000001084.12</td>\n",
       "      <td>ENSG00000001084.12</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid     source  type      start        end score strand phase  \\\n",
       "0  chr1  intersect  gene  169849631  169893959     .      -     .   \n",
       "1  chr1  intersect  gene  169794730  169854080     .      +     .   \n",
       "2  chr1  intersect  gene   27612064   27635277     .      -     .   \n",
       "3  chr1  intersect  gene  196651878  196747504     .      +     .   \n",
       "4  chr6  intersect  gene   53497341   53545129     .      -     .   \n",
       "\n",
       "              attributes              geneID  transID  \n",
       "0  ID=ENSG00000000457.13  ENSG00000000457.13      inf  \n",
       "1  ID=ENSG00000000460.16  ENSG00000000460.16      inf  \n",
       "2  ID=ENSG00000000938.12  ENSG00000000938.12      inf  \n",
       "3  ID=ENSG00000000971.15  ENSG00000000971.15      inf  \n",
       "4  ID=ENSG00000001084.12  ENSG00000001084.12      inf  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df[\"geneID\"]=gene_df.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "assert len(gene_df[gene_df[\"geneID\"].isnull()])==0,\"some IDs are missing from genes\"\n",
    "gene_df[\"transID\"]=np.inf\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "lower_df=lower_df.merge(full_df_tr[[\"transID\",\"geneID\"]],how=\"left\",left_on=\"transID\",right_on=\"transID\")\n",
    "\n",
    "dfM=pd.concat([gene_df[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]],\n",
    "        full_df_tr[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]],\n",
    "        lower_df[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]]],axis=0).reset_index(drop=True)\n",
    "dfM[\"type\"]=pd.Categorical(dfM[\"type\"],categories=[\"gene\",\"transcript\",\"exon\",\"CDS\"],ordered=True)\n",
    "dfM=dfM.sort_values(by=[\"seqid\",\"strand\",\"geneID\",\"transID\",\"type\"]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# lastly need to remove alternative scaffolds\n",
    "setChr=set(dfM['seqid'])\n",
    "alts = set(dfM[dfM[\"seqid\"].str.endswith(\"_alt\")][\"seqid\"])\n",
    "prims = [\"chr1\",\"chr2\",\"chr3\",\"chr4\",\"chr5\",\"chr6\",\"chr7\",\"chr8\",\"chr9\",\"chr10\",\"chr11\",\"chr12\",\"chr13\",\n",
    "         \"chr14\",\"chr15\",\"chr16\",\"chr17\",\"chr18\",\"chr19\",\"chr20\",\"chr21\",\"chr22\",\"chrX\",\"chrY\",\"chrM\"]\n",
    "rest = setChr - (alts.union(set(prims)))\n",
    "\n",
    "subdf = dfM[dfM[\"seqid\"].isin(set(prims).union(set(rest)))].reset_index(drop=True)\n",
    "subdf.to_csv(\"./prepAnnotation/refseq_vs_gencode_intersection.prot_lnc.no_alts.gff\",sep=\"\\t\",index=False,header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
