{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re\n",
    "\n",
    "gff3Cols=[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]\n",
    "chrMapCols=[\"name\",\"role\",\"molecule\",\"type\",\"genbank\",\"rel\",\"refseq\",\"unit\",\"seqLen\",\"ucsc\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we shall develop a method for merging two GFF files\n",
    "# Main idea - need to convert any arbitrary features into gene-transcript-exon features and discard anything unnecessary (such as CDS)\n",
    "#    - first we shall try to guess parent-child hierarchy\n",
    "#    - based on the hierarchy we can decide which entries to keep\n",
    "#    - alternatively, we can rely on exons, and assign \"transcript\" to their features and \"gene\" to the respective grandparents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>role</th>\n",
       "      <th>molecule</th>\n",
       "      <th>type</th>\n",
       "      <th>genbank</th>\n",
       "      <th>rel</th>\n",
       "      <th>refseq</th>\n",
       "      <th>unit</th>\n",
       "      <th>seqLen</th>\n",
       "      <th>ucsc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000663.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>248956422</td>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>2</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000664.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>242193529</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>3</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000665.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000003.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>198295559</td>\n",
       "      <td>chr3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>4</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000666.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>190214555</td>\n",
       "      <td>chr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>5</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000667.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>181538259</td>\n",
       "      <td>chr5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>6</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000668.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>170805979</td>\n",
       "      <td>chr6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>7</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>159345973</td>\n",
       "      <td>chr7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>8</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000670.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>145138636</td>\n",
       "      <td>chr8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>9</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000671.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>138394717</td>\n",
       "      <td>chr9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>10</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000672.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>133797422</td>\n",
       "      <td>chr10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>11</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000673.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>135086622</td>\n",
       "      <td>chr11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>12</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>133275309</td>\n",
       "      <td>chr12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>13</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000675.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>114364328</td>\n",
       "      <td>chr13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>14</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000676.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000014.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>107043718</td>\n",
       "      <td>chr14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>15</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000677.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>101991189</td>\n",
       "      <td>chr15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>16</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000678.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>90338345</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>17</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000679.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>83257441</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>18</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000680.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>80373285</td>\n",
       "      <td>chr18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000681.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>58617616</td>\n",
       "      <td>chr19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>20</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000682.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>64444167</td>\n",
       "      <td>chr20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>21</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000683.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>46709983</td>\n",
       "      <td>chr21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>22</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000684.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>50818468</td>\n",
       "      <td>chr22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>X</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>X</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000685.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>156040895</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Y</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>Y</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>CM000686.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>57227415</td>\n",
       "      <td>chrY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HSCHR1_CTG1_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270706.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187361.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>175055</td>\n",
       "      <td>chr1_KI270706v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HSCHR1_CTG2_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270707.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187362.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>32032</td>\n",
       "      <td>chr1_KI270707v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HSCHR1_CTG3_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270708.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187363.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>127682</td>\n",
       "      <td>chr1_KI270708v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HSCHR1_CTG4_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270709.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187364.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>66860</td>\n",
       "      <td>chr1_KI270709v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HSCHR1_CTG5_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270710.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187365.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>40176</td>\n",
       "      <td>chr1_KI270710v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HSCHR1_CTG6_UNLOCALIZED</td>\n",
       "      <td>unlocalized-scaffold</td>\n",
       "      <td>1</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270711.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187366.1</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>42210</td>\n",
       "      <td>chr1_KI270711v1_random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>HSCHR19LRC_PGF2_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL949753.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NW_003571061.2</td>\n",
       "      <td>ALT_REF_LOCI_8</td>\n",
       "      <td>796479</td>\n",
       "      <td>chr19_GL949753v2_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>HSCHR19_4_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270938.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187693.1</td>\n",
       "      <td>ALT_REF_LOCI_9</td>\n",
       "      <td>1066800</td>\n",
       "      <td>chr19_KI270938v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>HSCHR19KIR_FH15_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270882.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187636.1</td>\n",
       "      <td>ALT_REF_LOCI_10</td>\n",
       "      <td>248807</td>\n",
       "      <td>chr19_KI270882v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>HSCHR19KIR_G085_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270883.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187637.1</td>\n",
       "      <td>ALT_REF_LOCI_11</td>\n",
       "      <td>170399</td>\n",
       "      <td>chr19_KI270883v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>569</th>\n",
       "      <td>HSCHR19KIR_G085_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270884.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187638.1</td>\n",
       "      <td>ALT_REF_LOCI_12</td>\n",
       "      <td>157053</td>\n",
       "      <td>chr19_KI270884v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570</th>\n",
       "      <td>HSCHR19KIR_G248_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270885.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187639.1</td>\n",
       "      <td>ALT_REF_LOCI_13</td>\n",
       "      <td>171027</td>\n",
       "      <td>chr19_KI270885v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>HSCHR19KIR_G248_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270886.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187640.1</td>\n",
       "      <td>ALT_REF_LOCI_14</td>\n",
       "      <td>204239</td>\n",
       "      <td>chr19_KI270886v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>HSCHR19KIR_GRC212_AB_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270887.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187641.1</td>\n",
       "      <td>ALT_REF_LOCI_15</td>\n",
       "      <td>209512</td>\n",
       "      <td>chr19_KI270887v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>HSCHR19KIR_GRC212_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270888.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187642.1</td>\n",
       "      <td>ALT_REF_LOCI_16</td>\n",
       "      <td>155532</td>\n",
       "      <td>chr19_KI270888v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>HSCHR19KIR_LUCE_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270889.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187643.1</td>\n",
       "      <td>ALT_REF_LOCI_17</td>\n",
       "      <td>170698</td>\n",
       "      <td>chr19_KI270889v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270890.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187644.1</td>\n",
       "      <td>ALT_REF_LOCI_18</td>\n",
       "      <td>184499</td>\n",
       "      <td>chr19_KI270890v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>HSCHR19KIR_RSH_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270891.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187645.1</td>\n",
       "      <td>ALT_REF_LOCI_19</td>\n",
       "      <td>170680</td>\n",
       "      <td>chr19_KI270891v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>HSCHR19KIR_RSH_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270914.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187668.1</td>\n",
       "      <td>ALT_REF_LOCI_20</td>\n",
       "      <td>205194</td>\n",
       "      <td>chr19_KI270914v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578</th>\n",
       "      <td>HSCHR19KIR_T7526_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270915.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187669.1</td>\n",
       "      <td>ALT_REF_LOCI_21</td>\n",
       "      <td>170665</td>\n",
       "      <td>chr19_KI270915v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>HSCHR19KIR_T7526_BDEL_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270916.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187670.1</td>\n",
       "      <td>ALT_REF_LOCI_22</td>\n",
       "      <td>184516</td>\n",
       "      <td>chr19_KI270916v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>HSCHR19KIR_ABC08_A1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270917.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187671.1</td>\n",
       "      <td>ALT_REF_LOCI_23</td>\n",
       "      <td>190932</td>\n",
       "      <td>chr19_KI270917v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270918.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187672.1</td>\n",
       "      <td>ALT_REF_LOCI_24</td>\n",
       "      <td>123111</td>\n",
       "      <td>chr19_KI270918v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270919.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187673.1</td>\n",
       "      <td>ALT_REF_LOCI_25</td>\n",
       "      <td>170701</td>\n",
       "      <td>chr19_KI270919v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>HSCHR19KIR_FH05_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270920.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187674.1</td>\n",
       "      <td>ALT_REF_LOCI_26</td>\n",
       "      <td>198005</td>\n",
       "      <td>chr19_KI270920v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>HSCHR19KIR_FH05_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270921.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187675.1</td>\n",
       "      <td>ALT_REF_LOCI_27</td>\n",
       "      <td>282224</td>\n",
       "      <td>chr19_KI270921v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>HSCHR19KIR_FH06_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270922.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187676.1</td>\n",
       "      <td>ALT_REF_LOCI_28</td>\n",
       "      <td>187935</td>\n",
       "      <td>chr19_KI270922v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>HSCHR19KIR_FH06_BA1_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270923.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187677.1</td>\n",
       "      <td>ALT_REF_LOCI_29</td>\n",
       "      <td>189352</td>\n",
       "      <td>chr19_KI270923v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>HSCHR19KIR_FH08_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270929.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187683.1</td>\n",
       "      <td>ALT_REF_LOCI_30</td>\n",
       "      <td>186203</td>\n",
       "      <td>chr19_KI270929v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>HSCHR19KIR_FH08_BAX_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270930.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187684.1</td>\n",
       "      <td>ALT_REF_LOCI_31</td>\n",
       "      <td>200773</td>\n",
       "      <td>chr19_KI270930v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>HSCHR19KIR_FH13_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270931.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187685.1</td>\n",
       "      <td>ALT_REF_LOCI_32</td>\n",
       "      <td>170148</td>\n",
       "      <td>chr19_KI270931v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>HSCHR19KIR_FH13_BA2_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270932.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187686.1</td>\n",
       "      <td>ALT_REF_LOCI_33</td>\n",
       "      <td>215732</td>\n",
       "      <td>chr19_KI270932v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>HSCHR19KIR_FH15_A_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>KI270933.1</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_187687.1</td>\n",
       "      <td>ALT_REF_LOCI_34</td>\n",
       "      <td>170537</td>\n",
       "      <td>chr19_KI270933v1_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>HSCHR19KIR_RP5_B_HAP_CTG3_1</td>\n",
       "      <td>alt-scaffold</td>\n",
       "      <td>19</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>GL000209.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NT_113949.2</td>\n",
       "      <td>ALT_REF_LOCI_35</td>\n",
       "      <td>177381</td>\n",
       "      <td>chr19_GL000209v2_alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>MT</td>\n",
       "      <td>assembled-molecule</td>\n",
       "      <td>MT</td>\n",
       "      <td>Mitochondrion</td>\n",
       "      <td>J01415.2</td>\n",
       "      <td>=</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>non-nuclear</td>\n",
       "      <td>16569</td>\n",
       "      <td>chrM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>HSCHRUN_RANDOM_CTG29</td>\n",
       "      <td>unplaced-scaffold</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>KI270752.1</td>\n",
       "      <td>&lt;&gt;</td>\n",
       "      <td>na</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>27745</td>\n",
       "      <td>chrUn_KI270752v1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>595 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   name                  role molecule  \\\n",
       "0                                     1    assembled-molecule        1   \n",
       "1                                     2    assembled-molecule        2   \n",
       "2                                     3    assembled-molecule        3   \n",
       "3                                     4    assembled-molecule        4   \n",
       "4                                     5    assembled-molecule        5   \n",
       "5                                     6    assembled-molecule        6   \n",
       "6                                     7    assembled-molecule        7   \n",
       "7                                     8    assembled-molecule        8   \n",
       "8                                     9    assembled-molecule        9   \n",
       "9                                    10    assembled-molecule       10   \n",
       "10                                   11    assembled-molecule       11   \n",
       "11                                   12    assembled-molecule       12   \n",
       "12                                   13    assembled-molecule       13   \n",
       "13                                   14    assembled-molecule       14   \n",
       "14                                   15    assembled-molecule       15   \n",
       "15                                   16    assembled-molecule       16   \n",
       "16                                   17    assembled-molecule       17   \n",
       "17                                   18    assembled-molecule       18   \n",
       "18                                   19    assembled-molecule       19   \n",
       "19                                   20    assembled-molecule       20   \n",
       "20                                   21    assembled-molecule       21   \n",
       "21                                   22    assembled-molecule       22   \n",
       "22                                    X    assembled-molecule        X   \n",
       "23                                    Y    assembled-molecule        Y   \n",
       "24              HSCHR1_CTG1_UNLOCALIZED  unlocalized-scaffold        1   \n",
       "25              HSCHR1_CTG2_UNLOCALIZED  unlocalized-scaffold        1   \n",
       "26              HSCHR1_CTG3_UNLOCALIZED  unlocalized-scaffold        1   \n",
       "27              HSCHR1_CTG4_UNLOCALIZED  unlocalized-scaffold        1   \n",
       "28              HSCHR1_CTG5_UNLOCALIZED  unlocalized-scaffold        1   \n",
       "29              HSCHR1_CTG6_UNLOCALIZED  unlocalized-scaffold        1   \n",
       "..                                  ...                   ...      ...   \n",
       "565              HSCHR19LRC_PGF2_CTG3_1          alt-scaffold       19   \n",
       "566                    HSCHR19_4_CTG3_1          alt-scaffold       19   \n",
       "567        HSCHR19KIR_FH15_B_HAP_CTG3_1          alt-scaffold       19   \n",
       "568        HSCHR19KIR_G085_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "569      HSCHR19KIR_G085_BA1_HAP_CTG3_1          alt-scaffold       19   \n",
       "570        HSCHR19KIR_G248_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "571      HSCHR19KIR_G248_BA2_HAP_CTG3_1          alt-scaffold       19   \n",
       "572     HSCHR19KIR_GRC212_AB_HAP_CTG3_1          alt-scaffold       19   \n",
       "573    HSCHR19KIR_GRC212_BA1_HAP_CTG3_1          alt-scaffold       19   \n",
       "574        HSCHR19KIR_LUCE_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "575     HSCHR19KIR_LUCE_BDEL_HAP_CTG3_1          alt-scaffold       19   \n",
       "576         HSCHR19KIR_RSH_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "577       HSCHR19KIR_RSH_BA2_HAP_CTG3_1          alt-scaffold       19   \n",
       "578       HSCHR19KIR_T7526_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "579    HSCHR19KIR_T7526_BDEL_HAP_CTG3_1          alt-scaffold       19   \n",
       "580      HSCHR19KIR_ABC08_A1_HAP_CTG3_1          alt-scaffold       19   \n",
       "581  HSCHR19KIR_ABC08_AB_HAP_C_P_CTG3_1          alt-scaffold       19   \n",
       "582  HSCHR19KIR_ABC08_AB_HAP_T_P_CTG3_1          alt-scaffold       19   \n",
       "583        HSCHR19KIR_FH05_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "584        HSCHR19KIR_FH05_B_HAP_CTG3_1          alt-scaffold       19   \n",
       "585        HSCHR19KIR_FH06_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "586      HSCHR19KIR_FH06_BA1_HAP_CTG3_1          alt-scaffold       19   \n",
       "587        HSCHR19KIR_FH08_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "588      HSCHR19KIR_FH08_BAX_HAP_CTG3_1          alt-scaffold       19   \n",
       "589        HSCHR19KIR_FH13_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "590      HSCHR19KIR_FH13_BA2_HAP_CTG3_1          alt-scaffold       19   \n",
       "591        HSCHR19KIR_FH15_A_HAP_CTG3_1          alt-scaffold       19   \n",
       "592         HSCHR19KIR_RP5_B_HAP_CTG3_1          alt-scaffold       19   \n",
       "593                                  MT    assembled-molecule       MT   \n",
       "594                HSCHRUN_RANDOM_CTG29     unplaced-scaffold       na   \n",
       "\n",
       "              type     genbank rel          refseq              unit  \\\n",
       "0       Chromosome  CM000663.2   =    NC_000001.11  Primary Assembly   \n",
       "1       Chromosome  CM000664.2   =    NC_000002.12  Primary Assembly   \n",
       "2       Chromosome  CM000665.2   =    NC_000003.12  Primary Assembly   \n",
       "3       Chromosome  CM000666.2   =    NC_000004.12  Primary Assembly   \n",
       "4       Chromosome  CM000667.2   =    NC_000005.10  Primary Assembly   \n",
       "5       Chromosome  CM000668.2   =    NC_000006.12  Primary Assembly   \n",
       "6       Chromosome  CM000669.2   =    NC_000007.14  Primary Assembly   \n",
       "7       Chromosome  CM000670.2   =    NC_000008.11  Primary Assembly   \n",
       "8       Chromosome  CM000671.2   =    NC_000009.12  Primary Assembly   \n",
       "9       Chromosome  CM000672.2   =    NC_000010.11  Primary Assembly   \n",
       "10      Chromosome  CM000673.2   =    NC_000011.10  Primary Assembly   \n",
       "11      Chromosome  CM000674.2   =    NC_000012.12  Primary Assembly   \n",
       "12      Chromosome  CM000675.2   =    NC_000013.11  Primary Assembly   \n",
       "13      Chromosome  CM000676.2   =     NC_000014.9  Primary Assembly   \n",
       "14      Chromosome  CM000677.2   =    NC_000015.10  Primary Assembly   \n",
       "15      Chromosome  CM000678.2   =    NC_000016.10  Primary Assembly   \n",
       "16      Chromosome  CM000679.2   =    NC_000017.11  Primary Assembly   \n",
       "17      Chromosome  CM000680.2   =    NC_000018.10  Primary Assembly   \n",
       "18      Chromosome  CM000681.2   =    NC_000019.10  Primary Assembly   \n",
       "19      Chromosome  CM000682.2   =    NC_000020.11  Primary Assembly   \n",
       "20      Chromosome  CM000683.2   =     NC_000021.9  Primary Assembly   \n",
       "21      Chromosome  CM000684.2   =    NC_000022.11  Primary Assembly   \n",
       "22      Chromosome  CM000685.2   =    NC_000023.11  Primary Assembly   \n",
       "23      Chromosome  CM000686.2   =    NC_000024.10  Primary Assembly   \n",
       "24      Chromosome  KI270706.1   =     NT_187361.1  Primary Assembly   \n",
       "25      Chromosome  KI270707.1   =     NT_187362.1  Primary Assembly   \n",
       "26      Chromosome  KI270708.1   =     NT_187363.1  Primary Assembly   \n",
       "27      Chromosome  KI270709.1   =     NT_187364.1  Primary Assembly   \n",
       "28      Chromosome  KI270710.1   =     NT_187365.1  Primary Assembly   \n",
       "29      Chromosome  KI270711.1   =     NT_187366.1  Primary Assembly   \n",
       "..             ...         ...  ..             ...               ...   \n",
       "565     Chromosome  GL949753.2   =  NW_003571061.2    ALT_REF_LOCI_8   \n",
       "566     Chromosome  KI270938.1   =     NT_187693.1    ALT_REF_LOCI_9   \n",
       "567     Chromosome  KI270882.1   =     NT_187636.1   ALT_REF_LOCI_10   \n",
       "568     Chromosome  KI270883.1   =     NT_187637.1   ALT_REF_LOCI_11   \n",
       "569     Chromosome  KI270884.1   =     NT_187638.1   ALT_REF_LOCI_12   \n",
       "570     Chromosome  KI270885.1   =     NT_187639.1   ALT_REF_LOCI_13   \n",
       "571     Chromosome  KI270886.1   =     NT_187640.1   ALT_REF_LOCI_14   \n",
       "572     Chromosome  KI270887.1   =     NT_187641.1   ALT_REF_LOCI_15   \n",
       "573     Chromosome  KI270888.1   =     NT_187642.1   ALT_REF_LOCI_16   \n",
       "574     Chromosome  KI270889.1   =     NT_187643.1   ALT_REF_LOCI_17   \n",
       "575     Chromosome  KI270890.1   =     NT_187644.1   ALT_REF_LOCI_18   \n",
       "576     Chromosome  KI270891.1   =     NT_187645.1   ALT_REF_LOCI_19   \n",
       "577     Chromosome  KI270914.1   =     NT_187668.1   ALT_REF_LOCI_20   \n",
       "578     Chromosome  KI270915.1   =     NT_187669.1   ALT_REF_LOCI_21   \n",
       "579     Chromosome  KI270916.1   =     NT_187670.1   ALT_REF_LOCI_22   \n",
       "580     Chromosome  KI270917.1   =     NT_187671.1   ALT_REF_LOCI_23   \n",
       "581     Chromosome  KI270918.1   =     NT_187672.1   ALT_REF_LOCI_24   \n",
       "582     Chromosome  KI270919.1   =     NT_187673.1   ALT_REF_LOCI_25   \n",
       "583     Chromosome  KI270920.1   =     NT_187674.1   ALT_REF_LOCI_26   \n",
       "584     Chromosome  KI270921.1   =     NT_187675.1   ALT_REF_LOCI_27   \n",
       "585     Chromosome  KI270922.1   =     NT_187676.1   ALT_REF_LOCI_28   \n",
       "586     Chromosome  KI270923.1   =     NT_187677.1   ALT_REF_LOCI_29   \n",
       "587     Chromosome  KI270929.1   =     NT_187683.1   ALT_REF_LOCI_30   \n",
       "588     Chromosome  KI270930.1   =     NT_187684.1   ALT_REF_LOCI_31   \n",
       "589     Chromosome  KI270931.1   =     NT_187685.1   ALT_REF_LOCI_32   \n",
       "590     Chromosome  KI270932.1   =     NT_187686.1   ALT_REF_LOCI_33   \n",
       "591     Chromosome  KI270933.1   =     NT_187687.1   ALT_REF_LOCI_34   \n",
       "592     Chromosome  GL000209.2   =     NT_113949.2   ALT_REF_LOCI_35   \n",
       "593  Mitochondrion    J01415.2   =     NC_012920.1       non-nuclear   \n",
       "594             na  KI270752.1  <>              na  Primary Assembly   \n",
       "\n",
       "        seqLen                    ucsc  \n",
       "0    248956422                    chr1  \n",
       "1    242193529                    chr2  \n",
       "2    198295559                    chr3  \n",
       "3    190214555                    chr4  \n",
       "4    181538259                    chr5  \n",
       "5    170805979                    chr6  \n",
       "6    159345973                    chr7  \n",
       "7    145138636                    chr8  \n",
       "8    138394717                    chr9  \n",
       "9    133797422                   chr10  \n",
       "10   135086622                   chr11  \n",
       "11   133275309                   chr12  \n",
       "12   114364328                   chr13  \n",
       "13   107043718                   chr14  \n",
       "14   101991189                   chr15  \n",
       "15    90338345                   chr16  \n",
       "16    83257441                   chr17  \n",
       "17    80373285                   chr18  \n",
       "18    58617616                   chr19  \n",
       "19    64444167                   chr20  \n",
       "20    46709983                   chr21  \n",
       "21    50818468                   chr22  \n",
       "22   156040895                    chrX  \n",
       "23    57227415                    chrY  \n",
       "24      175055  chr1_KI270706v1_random  \n",
       "25       32032  chr1_KI270707v1_random  \n",
       "26      127682  chr1_KI270708v1_random  \n",
       "27       66860  chr1_KI270709v1_random  \n",
       "28       40176  chr1_KI270710v1_random  \n",
       "29       42210  chr1_KI270711v1_random  \n",
       "..         ...                     ...  \n",
       "565     796479    chr19_GL949753v2_alt  \n",
       "566    1066800    chr19_KI270938v1_alt  \n",
       "567     248807    chr19_KI270882v1_alt  \n",
       "568     170399    chr19_KI270883v1_alt  \n",
       "569     157053    chr19_KI270884v1_alt  \n",
       "570     171027    chr19_KI270885v1_alt  \n",
       "571     204239    chr19_KI270886v1_alt  \n",
       "572     209512    chr19_KI270887v1_alt  \n",
       "573     155532    chr19_KI270888v1_alt  \n",
       "574     170698    chr19_KI270889v1_alt  \n",
       "575     184499    chr19_KI270890v1_alt  \n",
       "576     170680    chr19_KI270891v1_alt  \n",
       "577     205194    chr19_KI270914v1_alt  \n",
       "578     170665    chr19_KI270915v1_alt  \n",
       "579     184516    chr19_KI270916v1_alt  \n",
       "580     190932    chr19_KI270917v1_alt  \n",
       "581     123111    chr19_KI270918v1_alt  \n",
       "582     170701    chr19_KI270919v1_alt  \n",
       "583     198005    chr19_KI270920v1_alt  \n",
       "584     282224    chr19_KI270921v1_alt  \n",
       "585     187935    chr19_KI270922v1_alt  \n",
       "586     189352    chr19_KI270923v1_alt  \n",
       "587     186203    chr19_KI270929v1_alt  \n",
       "588     200773    chr19_KI270930v1_alt  \n",
       "589     170148    chr19_KI270931v1_alt  \n",
       "590     215732    chr19_KI270932v1_alt  \n",
       "591     170537    chr19_KI270933v1_alt  \n",
       "592     177381    chr19_GL000209v2_alt  \n",
       "593      16569                    chrM  \n",
       "594      27745        chrUn_KI270752v1  \n",
       "\n",
       "[595 rows x 10 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrMap=pd.read_csv(\"/home/sparrow/genomicData/hg38/GCF_000001405.38_GRCh38.p12_assembly_report.txt\",sep=\"\\t\",names=chrMapCols,comment=\"#\")\n",
    "chrMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df=pd.read_csv(\"/home/sparrow/genomicData/hg38/annotations/hg38_p12_gencode_v29.gff3\",sep=\"\\t\",skiprows=7,names=gff3Cols)\n",
    "\n",
    "# df=df[df[\"seqid\"]==\"chr11\"].reset_index(drop=True)\n",
    "\n",
    "df[\"id\"]=df.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "dfE=df[df[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "dfE[\"transID\"]=dfE.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "transcriptSet=set(dfE[\"transID\"])\n",
    "dfT=df[df[\"id\"].isin(transcriptSet)].reset_index(drop=True)\n",
    "assert len(set(dfT[\"id\"]))==len(transcriptSet)\n",
    "dfT[\"geneID\"]=dfT.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "dfT[\"transID\"]=dfT[\"id\"]\n",
    "\n",
    "# here's the time to deal with inconsistencies\n",
    "# first remove any enties that do not have a parent - those do not conform to the expected structure that we define\n",
    "# however, this way we remove PSEUDOGENES, which may not be the desired effect - need to think more/discuss\n",
    "dfT=dfT[~(dfT[\"geneID\"].isnull())].reset_index(drop=True)\n",
    "dfE=dfE[dfE[\"transID\"].isin(dfT[\"transID\"])].reset_index(drop=True)\n",
    "\n",
    "# next it is time to iteratively, bring the parents to down\n",
    "# for instance, a feature that is defined as a transcript at this level (eg. miRNA) may have aother feature at this level as a parent\n",
    "while True:\n",
    "    dfI=dfT[[\"transID\",\"geneID\"]].merge(dfT[[\"transID\",\"geneID\"]],how=\"inner\",left_on=\"geneID\",right_on=\"transID\")\n",
    "    dfI=dfI[[\"transID_x\",\"geneID_y\"]]\n",
    "    dfT=dfT.merge(dfI,how=\"left\",left_on=\"transID\",right_on=\"transID_x\")\n",
    "    dfT[\"geneID\"]=np.where(dfT[\"geneID_y\"].isnull(),dfT[\"geneID\"],dfT[\"geneID_y\"])\n",
    "    if len(set(dfI[\"geneID_y\"]).intersection(set(dfT[\"transID\"])))==0:\n",
    "        break\n",
    "\n",
    "dfT[\"attributes\"]=dfT[\"attributes\"]+\";old_type=\"+dfT[\"type\"]\n",
    "dfT[\"type\"]=\"transcript\"\n",
    "geneSet=set(dfT[\"geneID\"])\n",
    "dfG=df[df[\"id\"].isin(geneSet)].reset_index(drop=True)\n",
    "dfG[\"attributes\"]=dfG[\"attributes\"]+\";old_type=\"+dfG[\"type\"]\n",
    "dfG[\"type\"]=\"gene\"\n",
    "assert len(set(dfG[\"id\"]))==len(geneSet)\n",
    "dfG[\"geneID\"]=dfG[\"id\"]\n",
    "dfG[\"transID\"]=np.inf\n",
    "\n",
    "dfE=dfE.merge(dfT[[\"transID\",\"geneID\"]],how=\"left\",left_on=\"transID\",right_on=\"transID\")\n",
    "assert len(dfE[dfE[\"geneID\"].isnull()])==0,\"incompatible features\"\n",
    "\n",
    "# now need to concatenate all these three dataframes\n",
    "dfM=pd.concat([dfG[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]],dfT[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]],dfE[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]]],axis=0).reset_index(drop=True)\n",
    "dfM[\"type\"]=pd.Categorical(dfM[\"type\"],categories=[\"gene\",\"transcript\",\"exon\"],ordered=True)\n",
    "dfM=dfM.sort_values(by=[\"geneID\",\"transID\",\"type\"]).reset_index(drop=True)\n",
    "# dfM[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]]\n",
    "del dfE\n",
    "del dfT\n",
    "del dfG\n",
    "# now need to convert the chromosome names to a standard (genbank)\n",
    "chrCurMap={}\n",
    "setNames=set(dfM[\"seqid\"])\n",
    "for i in setNames:\n",
    "    for j in [\"name\",\"molecule\",\"genbank\",\"refseq\",\"ucsc\"]:\n",
    "        if len(chrMap[chrMap[j]==i])>0:\n",
    "            chrCurMap[i]=chrMap[chrMap[j]==i][\"genbank\"].iloc[0]\n",
    "dfM.replace(chrCurMap,inplace=True)\n",
    "dfM[\"start\"]=dfM[\"start\"].astype(int)\n",
    "dfM[\"end\"]=dfM[\"end\"].astype(int)\n",
    "# dfM=dfM[gff3Cols]\n",
    "# some pseudogenes are still present\n",
    "\n",
    "# now we can proceed to merge annotations\n",
    "# dfM_g=dfM[dfM[\"type\"].isin([\"exon\"])][[\"seqid\",\"strand\",\"type\",\"start\",\"end\",\"transID\"]].reset_index(drop=True)\n",
    "# dfM_g=dfM_g.sort_values(by=[\"transID\",\"start\"]).reset_index(drop=True)\n",
    "# dfM_g[\"start\"]=dfM_g.start.astype(str)+\"+\"\n",
    "# dfM_g[\"end\"]=dfM_g.end.astype(str)+\"+\"\n",
    "# dfM_g=dfM_g.groupby(by=\"transID\").agg({'start':'sum','end':'sum','strand':'min','seqid':'min'})\n",
    "# dfM_g.reset_index(inplace=True)\n",
    "# dfM_g[\"chain\"]=dfM_g['seqid']+\":\"+dfM_g['strand']+'@'+dfM_g['start']+'-'+dfM_g['end']\n",
    "# dfM_g\n",
    "\n",
    "dfM_g=dfM[dfM[\"type\"].isin([\"exon\"])][[\"seqid\",\"strand\",\"type\",\"start\",\"end\",\"transID\"]].reset_index(drop=True)\n",
    "dfM_g=dfM_g.sort_values(by=[\"transID\",\"start\"]).reset_index(drop=True)\n",
    "dfM_g[\"ns\"]=dfM_g.start.astype(str).shift(-1)\n",
    "dfM_g[\"nid\"]=dfM_g.transID.shift(-1)\n",
    "dfM_g[\"ns\"]=np.where(dfM_g[\"transID\"]==dfM_g[\"nid\"],dfM_g[\"ns\"],\"***\")\n",
    "dfM_g[\"ns\"]=dfM_g.ns.astype(str)+\"+\"\n",
    "dfM_g[\"ends\"]=dfM_g.end.astype(str)+\"+\"\n",
    "dfM_g=dfM_g.groupby(by=\"transID\").agg({'start':'min','ns':'sum','ends':'sum','end':'max','strand':'min','seqid':'min'})\n",
    "dfM_g.reset_index(inplace=True)\n",
    "dfM_g.columns=['transID','start_min',\"ns\",'end','end_max','strand','seqid']\n",
    "dfM_g[\"ns\"]=dfM_g[\"ns\"].str.rstrip(\"***+\")\n",
    "dfM_g[\"end\"]=dfM_g[\"end\"].str.rstrip(\"***+\")\n",
    "dfM_g[\"chain\"]=dfM_g['seqid']+\":\"+dfM_g['strand']+'@'+dfM_g['end']+'-'+dfM_g['ns']\n",
    "dfM_g=dfM_g[['transID','start_min','end_max','chain']]\n",
    "\n",
    "del setNames\n",
    "del chrCurMap\n",
    "del geneSet\n",
    "del transcriptSet\n",
    "dfM=dfM[gff3Cols+[\"transID\"]]\n",
    "\n",
    "# now to remove duplicate intron chains\n",
    "# dfM_g[dfM_g.chain.isin(set(dfM_g[dfM_g[\"chain\"].duplicated()][\"chain\"]))].sort_values(by=\"chain\")\n",
    "dfM_g.drop_duplicates(\"chain\",keep='first',inplace=True)\n",
    "dfM_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df=pd.read_csv(\"/home/sparrow/genomicData/hg38/annotations/hg38_p12_refseq.gff\",sep=\"\\t\",skiprows=7,names=gff3Cols)\n",
    "\n",
    "# df=df[df[\"seqid\"]==\"NC_000011.10\"]\n",
    "\n",
    "df[\"id\"]=df.attributes.str.split(\"ID=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "dfE=df[df[\"type\"]==\"exon\"].reset_index(drop=True)\n",
    "dfE[\"transID\"]=dfE.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "transcriptSet=set(dfE[\"transID\"])\n",
    "dfT=df[df[\"id\"].isin(transcriptSet)].reset_index(drop=True)\n",
    "assert len(set(dfT[\"id\"]))==len(transcriptSet)\n",
    "dfT[\"geneID\"]=dfT.attributes.str.split(\"Parent=\",expand=True)[1].str.split(\";\",expand=True)[0]\n",
    "dfT[\"transID\"]=dfT[\"id\"]\n",
    "\n",
    "# here's the time to deal with inconsistencies\n",
    "# first remove any enties that do not have a parent - those do not conform to the expected structure that we define\n",
    "# however, this way we remove PSEUDOGENES, which may not be the desired effect - need to think more/discuss\n",
    "dfT=dfT[~(dfT[\"geneID\"].isnull())].reset_index(drop=True)\n",
    "dfE=dfE[dfE[\"transID\"].isin(dfT[\"transID\"])].reset_index(drop=True)\n",
    "\n",
    "# next it is time to iteratively, bring the parents to down\n",
    "# for instance, a feature that is defined as a transcript at this level (eg. miRNA) may have aother feature at this level as a parent\n",
    "while True:\n",
    "    dfI=dfT[[\"transID\",\"geneID\"]].merge(dfT[[\"transID\",\"geneID\"]],how=\"inner\",left_on=\"geneID\",right_on=\"transID\")\n",
    "    dfI=dfI[[\"transID_x\",\"geneID_y\"]]\n",
    "    dfT=dfT.merge(dfI,how=\"left\",left_on=\"transID\",right_on=\"transID_x\")\n",
    "    dfT[\"geneID\"]=np.where(dfT[\"geneID_y\"].isnull(),dfT[\"geneID\"],dfT[\"geneID_y\"])\n",
    "    if len(set(dfI[\"geneID_y\"]).intersection(set(dfT[\"transID\"])))==0:\n",
    "        break\n",
    "\n",
    "dfT[\"attributes\"]=dfT[\"attributes\"]+\";old_type=\"+dfT[\"type\"]\n",
    "dfT[\"type\"]=\"transcript\"\n",
    "geneSet=set(dfT[\"geneID\"])\n",
    "dfG=df[df[\"id\"].isin(geneSet)].reset_index(drop=True)\n",
    "del df\n",
    "dfG[\"attributes\"]=dfG[\"attributes\"]+\";old_type=\"+dfG[\"type\"]\n",
    "dfG[\"type\"]=\"gene\"\n",
    "assert len(set(dfG[\"id\"]))==len(geneSet)\n",
    "dfG[\"geneID\"]=dfG[\"id\"]\n",
    "dfG[\"transID\"]=np.inf\n",
    "\n",
    "dfE=dfE.merge(dfT[[\"transID\",\"geneID\"]],how=\"left\",left_on=\"transID\",right_on=\"transID\")\n",
    "assert len(dfE[dfE[\"geneID\"].isnull()])==0,\"incompatible features\"\n",
    "\n",
    "# now need to concatenate all these three dataframes\n",
    "dfM_2=pd.concat([dfG[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]],dfT[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]],dfE[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\",\"transID\",\"geneID\"]]],axis=0).reset_index(drop=True)\n",
    "dfM_2[\"type\"]=pd.Categorical(dfM_2[\"type\"],categories=[\"gene\",\"transcript\",\"exon\"],ordered=True)\n",
    "dfM_2=dfM_2.sort_values(by=[\"geneID\",\"transID\",\"type\"]).reset_index(drop=True)\n",
    "# dfM_2[[\"seqid\",\"source\",\"type\",\"start\",\"end\",\"score\",\"strand\",\"phase\",\"attributes\"]]\n",
    "dfE\n",
    "del dfT\n",
    "del dfG\n",
    "# now need to convert the chromosome names to a standard (genbank)\n",
    "chrCurMap={}\n",
    "setNames=set(dfM_2[\"seqid\"])\n",
    "for i in setNames:\n",
    "    for j in [\"name\",\"molecule\",\"genbank\",\"refseq\",\"ucsc\"]:\n",
    "        if len(chrMap[chrMap[j]==i])>0:\n",
    "            chrCurMap[i]=chrMap[chrMap[j]==i][\"genbank\"].iloc[0]\n",
    "dfM_2.replace(chrCurMap,inplace=True)\n",
    "dfM_2[\"start\"]=dfM_2[\"start\"].astype(int)\n",
    "dfM_2[\"end\"]=dfM_2[\"end\"].astype(int)\n",
    "# dfM_2=dfM_2[gff3Cols]\n",
    "# some pseudogenes are still present\n",
    "\n",
    "# now we can proceed to merge annotations\n",
    "# dfM_2_g=dfM_2[dfM_2[\"type\"].isin([\"exon\"])][[\"seqid\",\"strand\",\"type\",\"start\",\"end\",\"transID\"]].reset_index(drop=True)\n",
    "# dfM_2_g=dfM_2_g.sort_values(by=[\"transID\",\"start\"]).reset_index(drop=True)\n",
    "# dfM_2_g[\"start\"]=dfM_2_g.start.astype(str)+\"+\"\n",
    "# dfM_2_g[\"end\"]=dfM_2_g.end.astype(str)+\"+\"\n",
    "# dfM_2_g=dfM_2_g.groupby(by=\"transID\").agg({'start':'sum','end':'sum','strand':'min','seqid':'min'})\n",
    "# dfM_2_g.reset_index(inplace=True)\n",
    "# dfM_2_g[\"chain\"]=dfM_2_g['seqid']+\":\"+dfM_2_g['strand']+'@'+dfM_2_g['start']+'-'+dfM_2_g['end']\n",
    "# dfM_2_g\n",
    "\n",
    "dfM_2_g=dfM_2[dfM_2[\"type\"].isin([\"exon\"])][[\"seqid\",\"strand\",\"type\",\"start\",\"end\",\"transID\"]].reset_index(drop=True)\n",
    "dfM_2_g=dfM_2_g.sort_values(by=[\"transID\",\"start\"]).reset_index(drop=True)\n",
    "dfM_2_g[\"ns\"]=dfM_2_g.start.astype(str).shift(-1)\n",
    "dfM_2_g[\"nid\"]=dfM_2_g.transID.shift(-1)\n",
    "dfM_2_g[\"ns\"]=np.where(dfM_2_g[\"transID\"]==dfM_2_g[\"nid\"],dfM_2_g[\"ns\"],\"***\")\n",
    "dfM_2_g[\"ns\"]=dfM_2_g.ns.astype(str)+\"+\"\n",
    "dfM_2_g[\"ends\"]=dfM_2_g.end.astype(str)+\"+\"\n",
    "dfM_2_g=dfM_2_g.groupby(by=\"transID\").agg({'start':'min','ns':'sum','ends':'sum','end':'max','strand':'min','seqid':'min'})\n",
    "dfM_2_g.reset_index(inplace=True)\n",
    "dfM_2_g.columns=['transID','start_min',\"ns\",'end','end_max','strand','seqid']\n",
    "dfM_2_g[\"ns\"]=dfM_2_g[\"ns\"].str.rstrip(\"***+\")\n",
    "dfM_2_g[\"end\"]=dfM_2_g[\"end\"].str.rstrip(\"***+\")\n",
    "dfM_2_g[\"chain\"]=dfM_2_g['seqid']+\":\"+dfM_2_g['strand']+'@'+dfM_2_g['end']+'-'+dfM_2_g['ns']\n",
    "dfM_2_g=dfM_2_g[['transID','start_min','end_max','chain']]\n",
    "\n",
    "del setNames\n",
    "del chrCurMap\n",
    "del geneSet\n",
    "del transcriptSet\n",
    "dfM_2=dfM_2[gff3Cols+[\"transID\"]]\n",
    "\n",
    "# now to remove duplicate intron chains\n",
    "# dfM_2_g[dfM_2_g.chain.isin(set(dfM_2_g[dfM_2_g[\"chain\"].duplicated()][\"chain\"]))].sort_values(by=\"chain\")\n",
    "dfM_2_g.drop_duplicates(\"chain\",keep='first',inplace=True)\n",
    "dfM_2_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'dfM_g' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-72e971c6e4d2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# in doing so, we need to first identify all which overlap completely (that is having the sme intron chain,start and end)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mshared_chain\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdfM_g\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdfM_2_g\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"inner\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"chain\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;31m#[[\"transID_x\",'transID_y']]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0;32mdel\u001b[0m \u001b[0mdfM_g\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mdel\u001b[0m \u001b[0mdfM_2_g\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'dfM_g' is not defined"
     ]
    }
   ],
   "source": [
    "# now we can merge two annotations\n",
    "\n",
    "# in doing so, we need to first identify all which overlap completely (that is having the sme intron chain,start and end)\n",
    "shared_chain=dfM_g.merge(dfM_2_g,how=\"inner\",on=[\"chain\"])#[[\"transID_x\",'transID_y']]\n",
    "del dfM_g\n",
    "del dfM_2_g\n",
    "shared_all=shared_chain[(shared_chain['start_min_x']==shared_chain['start_min_y']) & (shared_chain['end_max_x']==shared_chain['end_max_y'])]\n",
    "# now we need to deal with those that share intron chain but have different start end\n",
    "# in these cases, the start or end or both exons need to be updated as well as the start or end or both of the transcripts\n",
    "sc=shared_chain[shared_chain[\"start_min_x\"]>shared_chain[\"start_min_y\"]]\n",
    "dfM=dfM.merge(sc,how=\"left\",left_on=\"transID\",right_on=\"transID_x\")\n",
    "dfM['start']=np.where(dfM['start']==dfM['start_min_x'],dfM['start_min_y'],dfM['start'])\n",
    "dfM=dfM[gff3Cols+[\"transID\"]]\n",
    "# second case\n",
    "sc=shared_chain[shared_chain[\"end_max_x\"]<shared_chain[\"end_max_y\"]]\n",
    "dfM=dfM.merge(sc,how=\"left\",left_on=\"transID\",right_on=\"transID_x\")\n",
    "\n",
    "dfM['end']=np.where(dfM['end']==dfM['end_max_x'],dfM['end_max_y'],dfM['end'])\n",
    "# now we only need to merge the rest of transcripts\n",
    "\n",
    "# right now there is no need to create new gene entries and standardize the parents\n",
    "# this is unnecessary since, we are using this annotation only to create a transcriptome fasta reference file\n",
    "# which by definition will only extract transcripts, and we only need to make sure there are no intron-chin duplicates\n",
    "dfM_2=dfM_2[~(dfM_2['transID'].isin(set(shared_chain[\"transID_x\"])))].reset_index(drop=True)\n",
    "# lastly, we shall build the final GFF by merging two annotations\n",
    "dfF=pd.concat([dfM[gff3Cols],dfM_2[gff3Cols]],axis=0)\n",
    "dfF.to_csv(\"./test.gff3\",sep='\\t',index=False,header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "# another thing: how should we handle pseudogenes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# at the end we shall provide add this interface to the trans2genome as an option to pass multiple annotation files"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
